


<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Index &mdash; polyCRACKER v1.1.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="#"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="polyCRACKER v1.1.1 documentation" href="index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> polyCRACKER
          

          
          </a>

          
            
            
              <div class="version">
                v1.1.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"></div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">polyCRACKER</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Index</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            

<h1 id="index">Index</h1>

<div class="genindex-jumpbox">
 <a href="#Symbols"><strong>Symbols</strong></a>
 | <a href="#P"><strong>P</strong></a>
 
</div>
<h2 id="Symbols">Symbols</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    --log

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-log">python-polyCRACKER.py-kcount_hist command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-old-log">python-polyCRACKER.py-kcount_hist_old command line option</a>
</li>
      </ul></li>
      <li>
    --shiny

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-karyotype-shiny">python-polyCRACKER.py-generate_karyotype command line option</a>
</li>
      </ul></li>
      <li>
    --version

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-version">python-polyCRACKER.py command line option</a>
</li>
      </ul></li>
      <li>
    -a, --account_name &lt;account_name&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-a">python-polyCRACKER.py-run_pipeline command line option</a>
</li>
      </ul></li>
      <li>
    -a, --anchor_file &lt;anchor_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-anchor2bed-a">python-polyCRACKER.py-anchor2bed command line option</a>
</li>
      </ul></li>
      <li>
    -all, --grab_all &lt;grab_all&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-all">python-polyCRACKER.py-cluster command line option</a>
</li>
      </ul></li>
      <li>
    -ax, --axes_off

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-ax">python-polyCRACKER.py-plotPositions command line option</a>
</li>
      </ul></li>
      <li>
    -b, --batch_script &lt;batch_script&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-b">python-polyCRACKER.py-run_pipeline command line option</a>
</li>
      </ul></li>
      <li>
    -b, --bb &lt;bb&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-b">python-polyCRACKER.py-blast2bed command line option</a>
</li>
      </ul></li>
      <li>
    -b, --bed_features_file &lt;bed_features_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-b">python-polyCRACKER.py-clusterGraph command line option</a>
</li>
      </ul></li>
      <li>
    -b, --bootstrap &lt;bootstrap&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-iqtree-b">python-polyCRACKER.py-run_iqtree command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-b">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -b, --final_output_label_bed &lt;final_output_label_bed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-b">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -b, --output_bed &lt;output_bed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-out-bed-to-circos-csv-b">python-polyCRACKER.py-out_bed_to_circos_csv command line option</a>
</li>
      </ul></li>
      <li>
    -b1, --bed1 &lt;bed1&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-b1">python-polyCRACKER.py-compareSubgenomes_progenitors_v_extracted command line option</a>
</li>
      </ul></li>
      <li>
    -b2, --bed2 &lt;bed2&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-b2">python-polyCRACKER.py-compareSubgenomes_progenitors_v_extracted command line option</a>
</li>
      </ul></li>
      <li>
    -b3, --bed3_directory &lt;bed3_directory&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-b3">python-polyCRACKER.py-generate_unionbed command line option</a>
</li>
      </ul></li>
      <li>
    -bb, --bb &lt;bb&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-bb">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -bo, --original_subgenomes_bed &lt;original_subgenomes_bed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-bo">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-bo">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -bt, --bootstrap

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-bt">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -c, --chunk_size &lt;chunk_size&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-c">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-c">python-polyCRACKER.py-splitFasta command line option</a>
</li>
      </ul></li>
      <li>
    -c, --cluster_analysis

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-c">python-polyCRACKER.py-final_stats command line option</a>
</li>
      </ul></li>
      <li>
    -c, --cluster_method &lt;cluster_method&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-c">python-polyCRACKER.py-cluster command line option</a>
</li>
      </ul></li>
      <li>
    -c, --colors_pickle &lt;colors_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-c">python-polyCRACKER.py-plotPositions command line option</a>
</li>
      </ul></li>
      <li>
    -c, --estimated_clusters &lt;estimated_clusters&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-merge-split-kmer-clusters-c">python-polyCRACKER.py-merge_split_kmer_clusters command line option</a>
</li>
      </ul></li>
      <li>
    -c, --input_labels &lt;input_labels&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-c">python-polyCRACKER.py-genomic_label_propagation command line option</a>
</li>
      </ul></li>
      <li>
    -cbed, --correspondence_bed &lt;correspondence_bed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-cbed">python-polyCRACKER.py-final_stats command line option</a>
</li>
      </ul></li>
      <li>
    -cs, --cluster_scaffold &lt;cluster_scaffold&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-cs">python-polyCRACKER.py-final_stats command line option</a>
</li>
      </ul></li>
      <li>
    -csv, --dist_mat &lt;dist_mat&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-csv">python-polyCRACKER.py-plot_distance_matrix command line option</a>
</li>
      </ul></li>
      <li>
    -csv, --input_csv &lt;input_csv&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-shiny2omic-csv">python-polyCRACKER.py-shiny2omic command line option</a>
</li>
      </ul></li>
      <li>
    -csv, --kmer_count_matrix &lt;kmer_count_matrix&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-csv">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-csv">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -csv, --kmer_rule_matrix &lt;kmer_rule_matrix&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genome-quality-assessment-csv">python-polyCRACKER.py-dash_genome_quality_assessment command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-csv">python-polyCRACKER.py-plot_rules_chromosomes command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-csv">python-polyCRACKER.py-return_dash_data_structures command line option</a>
</li>
      </ul></li>
      <li>
    -csv, --link_csv &lt;link_csv&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-link2color-csv">python-polyCRACKER.py-link2color command line option</a>
</li>
      </ul></li>
      <li>
    -csv, --multi_column_csv &lt;multi_column_csv&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-multicol2multifiles-csv">python-polyCRACKER.py-multicol2multifiles command line option</a>
</li>
      </ul></li>
      <li>
    -csv, --rule_csv &lt;rule_csv&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-csv">python-polyCRACKER.py-plot_rules command line option</a>
</li>
      </ul></li>
      <li>
    -d, --fasta_path &lt;fasta_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-d">python-polyCRACKER.py-splitFasta command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-writekmercount-d">python-polyCRACKER.py-writeKmerCount command line option</a>
</li>
      </ul></li>
      <li>
    -d, --kmercount_path &lt;kmercount_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-d">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
      </ul></li>
      <li>
    -d, --labels_dict &lt;labels_dict&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-scaffolds2colors-specified-d">python-polyCRACKER.py-scaffolds2colors_specified command line option</a>
</li>
      </ul></li>
      <li>
    -d, --n_dimensions &lt;n_dimensions&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-d">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-d">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-d">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-merge-split-kmer-clusters-d">python-polyCRACKER.py-merge_split_kmer_clusters command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-d">python-polyCRACKER.py-transform_plot command line option</a>
</li>
      </ul></li>
      <li>
    -d, --run_on_dir

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-hipmer-output-to-kcount-d">python-polyCRACKER.py-hipmer_output_to_kcount command line option</a>
</li>
      </ul></li>
      <li>
    -dash, --output_dash

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-dash">python-polyCRACKER.py-categorize_repeats command line option</a>
</li>
      </ul></li>
      <li>
    -dd, --diff_kmers_directory &lt;diff_kmers_directory&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-dd">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -dict, --comparisons_dict &lt;comparisons_dict&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-correct-kappa-dict">python-polyCRACKER.py-correct_kappa command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-dict">python-polyCRACKER.py-final_stats command line option</a>
</li>
      </ul></li>
      <li>
    -dk, --diff_kmer_threshold &lt;diff_kmer_threshold&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-dk">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -dk, --ratio_threshold &lt;ratio_threshold&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-dk">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
      </ul></li>
      <li>
    -ds, --diff_sample_rate &lt;diff_sample_rate&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-ds">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -dt, --diff_repeat_threshold &lt;diff_repeat_threshold&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-dt">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
      </ul></li>
      <li>
    -dv, --default_repeat_val &lt;default_repeat_val&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-dv">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
      </ul></li>
      <li>
    -env, --environment &lt;environment&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-env">python-polyCRACKER.py-run_pipeline command line option</a>
</li>
      </ul></li>
      <li>
    -ex, --external_call

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-ex">python-polyCRACKER.py-blast2bed command line option</a>
</li>
      </ul></li>
      <li>
    -f, --fasta_in &lt;fasta_in&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-iqtree-f">python-polyCRACKER.py-run_iqtree command line option</a>
</li>
      </ul></li>
      <li>
    -f, --fasta_path &lt;fasta_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-f">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
      </ul></li>
      <li>
    -f, --feature_space &lt;feature_space&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-f">python-polyCRACKER.py-plot_distance_matrix command line option</a>
</li>
      </ul></li>
      <li>
    -f, --file &lt;file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-f">python-polyCRACKER.py-cluster command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-f">python-polyCRACKER.py-find_best_cluster_parameters command line option</a>
</li>
      </ul></li>
      <li>
    -f, --folder_delimiter &lt;folder_delimiter&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-f">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
      </ul></li>
      <li>
    -fai, --fai_file &lt;fai_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-fai">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-out-bed-fai">python-polyCRACKER.py-generate_out_bed command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-fai">python-polyCRACKER.py-generate_unionbed command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-get-density-fai">python-polyCRACKER.py-get_density command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-out-bed-to-circos-csv-fai">python-polyCRACKER.py-out_bed_to_circos_csv command line option</a>
</li>
      </ul></li>
      <li>
    -fi, --fasta_in &lt;fasta_in&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-extract-sequences-fi">python-polyCRACKER.py-extract_sequences command line option</a>
</li>
      </ul></li>
      <li>
    -fi, --input_fasta &lt;input_fasta&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-denovo-repeats-fi">python-polyCRACKER.py-find_denovo_repeats command line option</a>
</li>
      </ul></li>
      <li>
    -fo, --output_fname &lt;output_fname&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-out-bed-fo">python-polyCRACKER.py-generate_out_bed command line option</a>
</li>
      </ul></li>
      <li>
    -fout, --outputfname &lt;outputfname&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-get-density-fout">python-polyCRACKER.py-get_density command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-fout">python-polyCRACKER.py-progenitorMapping command line option</a>
</li>
      </ul></li>
      <li>
    -fp, --fasta_path &lt;fasta_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-fp">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -ft, --fast_tree

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-ft">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
      </ul></li>
      <li>
    -g, --gamma &lt;gamma&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-g">python-polyCRACKER.py-cluster command line option</a>
</li>
      </ul></li>
      <li>
    -g, --generations_number &lt;generations_number&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-g">python-polyCRACKER.py-find_best_cluster_parameters command line option</a>
</li>
      </ul></li>
      <li>
    -g, --genome &lt;genome&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-g">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
      </ul></li>
      <li>
    -g, --genome_name &lt;genome_name&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-g">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -g, --pre_scaffolded_genomes &lt;pre_scaffolded_genomes&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genome-quality-assessment-g">python-polyCRACKER.py-dash_genome_quality_assessment command line option</a>
</li>
      </ul></li>
      <li>
    -gc, --gene_crossover_prob &lt;gene_crossover_prob&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-gc">python-polyCRACKER.py-find_best_cluster_parameters command line option</a>
</li>
      </ul></li>
      <li>
    -gff, --gff_file &lt;gff_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-get-density-gff">python-polyCRACKER.py-get_density command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-gff">python-polyCRACKER.py-progenitorMapping command line option</a>
</li>
      </ul></li>
      <li>
    -gff, --repeat_gff &lt;repeat_gff&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-gff">python-polyCRACKER.py-categorize_repeats command line option</a>
</li>
      </ul></li>
      <li>
    -gm, --gene_mutation_prob &lt;gene_mutation_prob&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-gm">python-polyCRACKER.py-find_best_cluster_parameters command line option</a>
</li>
      </ul></li>
      <li>
    -go, --original_fasta &lt;original_fasta&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-go">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -go, --original_genome &lt;original_genome&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-avg-distance-between-diff-kmers-go">python-polyCRACKER.py-avg_distance_between_diff_kmers command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-go">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-go">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-karyotype-go">python-polyCRACKER.py-generate_karyotype command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-go">python-polyCRACKER.py-plot_rules_chromosomes command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-go">python-polyCRACKER.py-progenitorMapping command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-go">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -gs, --genome_split_name &lt;genome_split_name&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-gs">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
      </ul></li>
      <li>
    -hb, --high_bool &lt;high_bool&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-hb">python-polyCRACKER.py-kmer2Fasta command line option</a>
</li>
      </ul></li>
      <li>
    -hc, --kmer_high_count &lt;kmer_high_count&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-hc">python-polyCRACKER.py-kmer2Fasta command line option</a>
</li>
      </ul></li>
      <li>
    -i, --bed_file &lt;bed_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bed2scaffolds-pickle-i">python-polyCRACKER.py-bed2scaffolds_pickle command line option</a>
</li>
      </ul></li>
      <li>
    -i, --blast_file &lt;blast_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-i">python-polyCRACKER.py-blast2bed command line option</a>
</li>
      </ul></li>
      <li>
    -i, --fasta_file &lt;fasta_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-i">python-polyCRACKER.py-splitFasta command line option</a>
</li>
      </ul></li>
      <li>
    -i, --hipmer_input &lt;hipmer_input&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-hipmer-output-to-kcount-i">python-polyCRACKER.py-hipmer_output_to_kcount command line option</a>
</li>
      </ul></li>
      <li>
    -i, --input_bed_file &lt;input_bed_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-i">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
      </ul></li>
      <li>
    -i, --input_dict &lt;input_dict&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-i">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
      </ul></li>
      <li>
    -i, --input_fasta &lt;input_fasta&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-i">python-polyCRACKER.py-progenitorMapping command line option</a>
</li>
      </ul></li>
      <li>
    -i, --input_files &lt;input_files&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-i">python-polyCRACKER.py-generate_genome_signatures command line option</a>
</li>
      </ul></li>
      <li>
    -i, --iteration &lt;iteration&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-i">python-polyCRACKER.py-clusterGraph command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-i">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -i, --iterations &lt;iterations&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-i">python-polyCRACKER.py-plotPositions command line option</a>
</li>
      </ul></li>
      <li>
    -id, --input_dir &lt;input_dir&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-id">python-polyCRACKER.py-convert_subgenome_output_to_pickle command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-out-bed-id">python-polyCRACKER.py-generate_out_bed command line option</a>
</li>
      </ul></li>
      <li>
    -im, --input_merged_bed_file &lt;input_merged_bed_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-im">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
      </ul></li>
      <li>
    -J, --job_name &lt;job_name&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-j">python-polyCRACKER.py-run_pipeline command line option</a>
</li>
      </ul></li>
      <li>
    -j, --n_jobs &lt;n_jobs&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-j">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-j">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-j">python-polyCRACKER.py-plot_distance_matrix command line option</a>
</li>
      </ul></li>
      <li>
    -k, --kcount_directory &lt;kcount_directory&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-k">python-polyCRACKER.py-kcount_hist command line option</a>
</li>
      </ul></li>
      <li>
    -k, --kcount_file &lt;kcount_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-old-k">python-polyCRACKER.py-kcount_hist_old command line option</a>
</li>
      </ul></li>
      <li>
    -k, --kernel &lt;kernel&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-k">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-k">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-k">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-k">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -k, --kmeans_clusters &lt;kmeans_clusters&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-k">python-polyCRACKER.py-plot_distance_matrix command line option</a>
</li>
      </ul></li>
      <li>
    -k, --kmer_fasta &lt;kmer_fasta&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-k">python-polyCRACKER.py-blast_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -k, --kmercount_path &lt;kmercount_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-k">python-polyCRACKER.py-kmer2Fasta command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-writekmercount-k">python-polyCRACKER.py-writeKmerCount command line option</a>
</li>
      </ul></li>
      <li>
    -k, --output_kmer_file &lt;output_kmer_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-k">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
      </ul></li>
      <li>
    -kl, --kmer_length &lt;kmer_length&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-kl">python-polyCRACKER.py-blast_kmers command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-kl">python-polyCRACKER.py-generate_genome_signatures command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-kl">python-polyCRACKER.py-progenitorMapping command line option</a>
</li>
      </ul></li>
      <li>
    -kl, --kmer_lengths &lt;kmer_lengths&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-kl">python-polyCRACKER.py-kcount_hist command line option</a>
</li>
      </ul></li>
      <li>
    -kl, --kmer_low_count &lt;kmer_low_count&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-kl">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -kl, --kmer_pca_labels &lt;kmer_pca_labels&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-kl">python-polyCRACKER.py-return_dash_data_structures command line option</a>
</li>
      </ul></li>
      <li>
    -kp, --kmers_pickle &lt;kmers_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-kp">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -kr, --kmer_rules_pickle &lt;kmer_rules_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-kr">python-polyCRACKER.py-return_dash_data_structures command line option</a>
</li>
      </ul></li>
      <li>
    -kv, --default_kcount_value &lt;default_kcount_value&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-kv">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
      </ul></li>
      <li>
    -kv, --default_kmercount_value &lt;default_kmercount_value&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-kv">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -l, --kmer_length &lt;kmer_length&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-l">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-l">python-polyCRACKER.py-categorize_repeats command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-l">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-writekmercount-l">python-polyCRACKER.py-writeKmerCount command line option</a>
</li>
      </ul></li>
      <li>
    -l, --kmer_list_path &lt;kmer_list_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-l">python-polyCRACKER.py-polyploid_diff_kmer_comparison command line option</a>
</li>
      </ul></li>
      <li>
    -l, --layout &lt;layout&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-l">python-polyCRACKER.py-clusterGraph command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-l">python-polyCRACKER.py-plotPositions command line option</a>
</li>
      </ul></li>
      <li>
    -l, --list_of_clusters &lt;list_of_clusters&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-exchange-l">python-polyCRACKER.py-cluster_exchange command line option</a>
</li>
      </ul></li>
      <li>
    -l, --low_count &lt;low_count&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-l">python-polyCRACKER.py-kcount_hist command line option</a>
</li>
      </ul></li>
      <li>
    -l, --low_memory &lt;low_memory&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-l">python-polyCRACKER.py-blast2bed command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-l">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-l">python-polyCRACKER.py-splitFasta command line option</a>
</li>
      </ul></li>
      <li>
    -lc, --kmer_low_count &lt;kmer_low_count&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-lc">python-polyCRACKER.py-kmer2Fasta command line option</a>
</li>
      </ul></li>
      <li>
    -low, --low_counts &lt;low_counts&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-low">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-low">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -m, --blast_mem &lt;blast_mem&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-m">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-m">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-m">python-polyCRACKER.py-blast_kmers command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-m">python-polyCRACKER.py-categorize_repeats command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genome-quality-assessment-m">python-polyCRACKER.py-dash_genome_quality_assessment command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-m">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-m">python-polyCRACKER.py-plot_rules_chromosomes command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-m">python-polyCRACKER.py-progenitorMapping command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-m">python-polyCRACKER.py-return_dash_data_structures command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-m">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-writekmercount-m">python-polyCRACKER.py-writeKmerCount command line option</a>
</li>
      </ul></li>
      <li>
    -m, --memory &lt;memory&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-m">python-polyCRACKER.py-run_pipeline command line option</a>
</li>
      </ul></li>
      <li>
    -m, --metric &lt;metric&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-m">python-polyCRACKER.py-cluster command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-m">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-m">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-m">python-polyCRACKER.py-transform_plot command line option</a>
</li>
      </ul></li>
      <li>
    -m, --min_chunk_size &lt;min_chunk_size&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-m">python-polyCRACKER.py-splitFasta command line option</a>
</li>
      </ul></li>
      <li>
    -m, --model &lt;model&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-iqtree-m">python-polyCRACKER.py-run_iqtree command line option</a>
</li>
      </ul></li>
      <li>
    -max, --max_val &lt;max_val&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-max">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
      </ul></li>
      <li>
    -mc, --min_chunk_size &lt;min_chunk_size&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-mc">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
      </ul></li>
      <li>
    -mf, --multi_fasta

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-mf">python-polyCRACKER.py-generate_genome_signatures command line option</a>
</li>
      </ul></li>
      <li>
    -min, --min_count &lt;min_count&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-min">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    -min, --min_number_diff_kmers &lt;min_number_diff_kmers&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-min">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -min_c, --min_cluster_size &lt;min_cluster_size&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-c">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-min-c">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -min_n, --min_number_clusters &lt;min_number_clusters&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-n">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-min-n">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -min_s, --min_samples &lt;min_samples&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-s">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-min-s">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -motif, --grab_motif

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-motif">python-polyCRACKER.py-categorize_repeats command line option</a>
</li>
      </ul></li>
      <li>
    -mst, --min_span_tree &lt;min_span_tree&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-mst">python-polyCRACKER.py-cluster command line option</a>
</li>
      </ul></li>
      <li>
    -n, --n_chromosomes &lt;n_chromosomes&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-n">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-n">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -n, --newick_in &lt;newick_in&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-color-trees-n">python-polyCRACKER.py-color_trees command line option</a>
</li>
      </ul></li>
      <li>
    -n, --nextflow_config &lt;nextflow_config&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-update-nextflow-config-n">python-polyCRACKER.py-update_nextflow_config command line option</a>
</li>
      </ul></li>
      <li>
    -n, --no_options

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-no-options">python-polyCRACKER.py-run_pipeline command line option</a>
</li>
      </ul></li>
      <li>
    -N, --nodes &lt;nodes&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-n">python-polyCRACKER.py-run_pipeline command line option</a>
</li>
      </ul></li>
      <li>
    -n, --note &lt;note&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-n">python-polyCRACKER.py-final_stats command line option</a>
</li>
      </ul></li>
      <li>
    -n, --number_chromosomes &lt;number_chromosomes&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-unionbed-n">python-polyCRACKER.py-plot_unionbed command line option</a>
</li>
      </ul></li>
      <li>
    -n, -n_rules &lt;n_rules&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-rules2-n">python-polyCRACKER.py-find_rules2 command line option</a>
</li>
      </ul></li>
      <li>
    -ndb, --no_diff_blast

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-ndb">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -nj, --no_jaccard_clustering

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-nj">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -nn, --n_neighbors &lt;n_neighbors&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-nn">python-polyCRACKER.py-cluster command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-nn">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-nn">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -nnk, --n_neighbors_kmers &lt;n_neighbors_kmers&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-nnk">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -nns, --n_neighbors_scaffolds &lt;n_neighbors_scaffolds&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-nns">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -np, --no_run_phylogeny

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-np">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -npy, --differential_repeat_subgenome_matrix &lt;differential_TE_subgenome_matrix&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-differential-te-histogram-npy">python-polyCRACKER.py-differential_TE_histogram command line option</a>
</li>
      </ul></li>
      <li>
    -npy, --kmer_pca_data &lt;kmer_pca_data&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-npy">python-polyCRACKER.py-return_dash_data_structures command line option</a>
</li>
      </ul></li>
      <li>
    -npy, --positions &lt;positions&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-spectral-embed-plot-npy">python-polyCRACKER.py-spectral_embed_plot command line option</a>
</li>
      </ul></li>
      <li>
    -npy, --positions_npy &lt;positions_npy&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-npy">python-polyCRACKER.py-final_stats command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-npy">python-polyCRACKER.py-plotPositions command line option</a>
</li>
      </ul></li>
      <li>
    -npy, --transformed_data &lt;transformed_data&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-npy">python-polyCRACKER.py-genomic_label_propagation command line option</a>
</li>
      </ul></li>
      <li>
    -npz, --graph_file &lt;graph_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-npz">python-polyCRACKER.py-plotPositions command line option</a>
</li>
      </ul></li>
      <li>
    -npz, --input_matrix &lt;input_matrix&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-convert-mat2r-npz">python-polyCRACKER.py-convert_mat2r command line option</a>
</li>
      </ul></li>
      <li>
    -npz, --output_sparse_matrix &lt;output_sparse_matrix&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-npz">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
      </ul></li>
      <li>
    -npz, --sparse_kmer_matrix &lt;sparse_kmer_matrix&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-npz">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -ns, --no_sam2matrix

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-ns">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -nt, --n_threads &lt;n_threads&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-iqtree-nt">python-polyCRACKER.py-run_iqtree command line option</a>
</li>
      </ul></li>
      <li>
    -nt, --n_top_repeats &lt;n_top_repeats&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-nt">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
      </ul></li>
      <li>
    -o, --kcount_output &lt;kcount_output&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-hipmer-output-to-kcount-o">python-polyCRACKER.py-hipmer_output_to_kcount command line option</a>
</li>
      </ul></li>
      <li>
    -o, --original &lt;original&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-out-bed-o">python-polyCRACKER.py-generate_out_bed command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-o">python-polyCRACKER.py-generate_unionbed command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-o">python-polyCRACKER.py-progenitorMapping command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-o">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -o, --out_file &lt;out_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-o">python-polyCRACKER.py-kcount_hist command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-old-o">python-polyCRACKER.py-kcount_hist_old command line option</a>
</li>
      </ul></li>
      <li>
    -o, --output_bed_file &lt;output_bed_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-o">python-polyCRACKER.py-blast2bed command line option</a>
</li>
      </ul></li>
      <li>
    -o, --output_dir &lt;output_dir&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-o">python-polyCRACKER.py-return_dash_data_structures command line option</a>
</li>
      </ul></li>
      <li>
    -o, --output_file &lt;output_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-o">python-polyCRACKER.py-blast_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -o, --output_file_prefix &lt;output_file_prefix&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-o">python-polyCRACKER.py-generate_genome_signatures command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-o">python-polyCRACKER.py-plot_distance_matrix command line option</a>
</li>
      </ul></li>
      <li>
    -o, --output_fname &lt;output_fname&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-o">python-polyCRACKER.py-plotPositions command line option</a>
</li>
      </ul></li>
      <li>
    -o, --output_matrix &lt;output_matrix&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-unionbed2matrix-o">python-polyCRACKER.py-unionbed2matrix command line option</a>
</li>
      </ul></li>
      <li>
    -o, --output_pickle &lt;output_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bed2scaffolds-pickle-o">python-polyCRACKER.py-bed2scaffolds_pickle command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-o">python-polyCRACKER.py-convert_subgenome_output_to_pickle command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-o">python-polyCRACKER.py-genomic_label_propagation command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-scaffolds2colors-specified-o">python-polyCRACKER.py-scaffolds2colors_specified command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-species-comparison-scaffold2colors-o">python-polyCRACKER.py-species_comparison_scaffold2colors command line option</a>
</li>
      </ul></li>
      <li>
    -od, --out_dir &lt;out_dir&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-od">python-polyCRACKER.py-clusterGraph command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-od">python-polyCRACKER.py-compareSubgenomes_progenitors_v_extracted command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-denovo-repeats-od">python-polyCRACKER.py-find_denovo_repeats command line option</a>
</li>
      </ul></li>
      <li>
    -od, --out_folder &lt;out_folder&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-unionbed-od">python-polyCRACKER.py-plot_unionbed command line option</a>
</li>
      </ul></li>
      <li>
    -om, --output_merged_bed_file &lt;output_merged_bed_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-om">python-polyCRACKER.py-blast2bed command line option</a>
</li>
      </ul></li>
      <li>
    -op, --original_progenitors &lt;original_progenitors&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-op">python-polyCRACKER.py-final_stats command line option</a>
</li>
      </ul></li>
      <li>
    -os, --original_subgenome_path &lt;original_subgenome_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-os">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -p, --analysis_path &lt;analysis_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-p">python-polyCRACKER.py-polyploid_diff_kmer_comparison command line option</a>
</li>
      </ul></li>
      <li>
    -p, --fasta_path &lt;fasta_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-p">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -p, --final_output_label_pickle &lt;final_output_label_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-p">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -p, --labels_pickle &lt;labels_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-p">python-polyCRACKER.py-plotPositions command line option</a>
</li>
      </ul></li>
      <li>
    -p, --polycracker_config &lt;polycracker_config&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-update-nextflow-config-p">python-polyCRACKER.py-update_nextflow_config command line option</a>
</li>
      </ul></li>
      <li>
    -p, --population_size &lt;population_size&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-p">python-polyCRACKER.py-find_best_cluster_parameters command line option</a>
</li>
      </ul></li>
      <li>
    -p, --positions_npy &lt;positions_npy&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-p">python-polyCRACKER.py-clusterGraph command line option</a>
</li>
      </ul></li>
      <li>
    -p, --prefilter &lt;prefilter&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-p">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-p">python-polyCRACKER.py-splitFasta command line option</a>
</li>
      </ul></li>
      <li>
    -p, --progenitor_fasta_folder &lt;progenitor_fasta_folder&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-p">python-polyCRACKER.py-progenitorMapping command line option</a>
</li>
      </ul></li>
      <li>
    -p, --progenitors, -s, --species

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-p">python-polyCRACKER.py-final_stats command line option</a>
</li>
      </ul></li>
      <li>
    -p0, --scaffolds_pickle &lt;scaffolds_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-correct-kappa-p0">python-polyCRACKER.py-correct_kappa command line option</a>
</li>
      </ul></li>
      <li>
    -p1, --polycracker_pickle &lt;polycracker_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-correct-kappa-p1">python-polyCRACKER.py-correct_kappa command line option</a>
</li>
      </ul></li>
      <li>
    -p2, --progenitors_pickle &lt;progenitors_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-correct-kappa-p2">python-polyCRACKER.py-correct_kappa command line option</a>
</li>
      </ul></li>
      <li>
    -pbed, --polycracker_bed &lt;polycracker_bed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-pbed">python-polyCRACKER.py-final_stats command line option</a>
</li>
      </ul></li>
      <li>
    -pf, --progenitor_fastas &lt;progenitor_fastas&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-pf">python-polyCRACKER.py-progenitorMapping command line option</a>
</li>
      </ul></li>
      <li>
    -pm, --perfect_mode &lt;perfect_mode&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-pm">python-polyCRACKER.py-blast_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -q, --qbed &lt;qbed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-anchor2bed-q">python-polyCRACKER.py-anchor2bed command line option</a>
</li>
      </ul></li>
      <li>
    -r, --all_family_subclass_pickle &lt;all_family_subclass_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-r">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -r, --kcount_range &lt;kcount_range&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-r">python-polyCRACKER.py-kcount_hist command line option</a>
</li>
      </ul></li>
      <li>
    -r, --radius &lt;radius&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-r">python-polyCRACKER.py-genomic_label_propagation command line option</a>
</li>
      </ul></li>
      <li>
    -r, --reduction_method &lt;reduction_method&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-r">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-r">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -r, --reference_genome &lt;reference_genome&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-r">python-polyCRACKER.py-blast_kmers command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-r">python-polyCRACKER.py-return_dash_data_structures command line option</a>
</li>
      </ul></li>
      <li>
    -r, --remove_non_chunk &lt;remove_non_chunk&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-r">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-r">python-polyCRACKER.py-splitFasta command line option</a>
</li>
      </ul></li>
      <li>
    -r, --repeat_masked_gff &lt;repeat_masked_gff&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-r">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -r, --rules_kmers &lt;rules_kmers&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-r">python-polyCRACKER.py-categorize_repeats command line option</a>
</li>
      </ul></li>
      <li>
    -r, --run_final &lt;run_final&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-r">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -rd, --recover_dir &lt;recover_dir&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-denovo-repeats-rd">python-polyCRACKER.py-find_denovo_repeats command line option</a>
</li>
      </ul></li>
      <li>
    -rd, --top_repeat_dict_pickle &lt;top_repeat_dict_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-color-trees-rd">python-polyCRACKER.py-color_trees command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-rd">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -re, --all_repeat_elements &lt;all_repeat_elements&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-re">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -rf, --reference_fasta &lt;reference_fasta&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-txt2fasta-rf">python-polyCRACKER.py-txt2fasta command line option</a>
</li>
      </ul></li>
      <li>
    -rf, --repeat_fasta &lt;repeat_fasta&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-rf">python-polyCRACKER.py-categorize_repeats command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-rf">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -rf, --subclass_repeat_fasta &lt;subclass_repeat_fasta&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-send-repeats-rf">python-polyCRACKER.py-send_repeats command line option</a>
</li>
      </ul></li>
      <li>
    -rg, --repeat_gff &lt;repeat_gff&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-rg">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
      </ul></li>
      <li>
    -rp, --repeat_pickle &lt;repeat_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-rp">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
      </ul></li>
      <li>
    -rp, --top_repeat_list_pickle &lt;top_repeat_list_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-rp">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -rs, --class_subclass_pickle &lt;class_subclass_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-rs">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -s, --graph_sampling_speed &lt;graph_sampling_speed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-s">python-polyCRACKER.py-plotPositions command line option</a>
</li>
      </ul></li>
      <li>
    -s, --n_clusters &lt;n_clusters&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-s">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
      </ul></li>
      <li>
    -s, --n_subgenomes &lt;n_subgenomes&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-s">python-polyCRACKER.py-cluster command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-n-subgenomes">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-s">python-polyCRACKER.py-find_best_cluster_parameters command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-s">python-polyCRACKER.py-transform_plot command line option</a>
</li>
      </ul></li>
      <li>
    -s, --sample_speed &lt;sample_speed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-s">python-polyCRACKER.py-kcount_hist command line option</a>
</li>
      </ul></li>
      <li>
    -s, --sampling_sensitivity &lt;sampling_sensitivity&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-s">python-polyCRACKER.py-kmer2Fasta command line option</a>
</li>
      </ul></li>
      <li>
    -s, --sbed &lt;sbed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-anchor2bed-s">python-polyCRACKER.py-anchor2bed command line option</a>
</li>
      </ul></li>
      <li>
    -s, --scaffolds_file &lt;scaffolds_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-s">python-polyCRACKER.py-compareSubgenomes_progenitors_v_extracted command line option</a>
</li>
      </ul></li>
      <li>
    -s, --scaffolds_pickle &lt;scaffolds_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-s">python-polyCRACKER.py-convert_subgenome_output_to_pickle command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-s">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-s">python-polyCRACKER.py-genomic_label_propagation command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-scaffolds2colors-specified-s">python-polyCRACKER.py-scaffolds2colors_specified command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-species-comparison-scaffold2colors-s">python-polyCRACKER.py-species_comparison_scaffold2colors command line option</a>
</li>
      </ul></li>
      <li>
    -s, --scaled &lt;scaled&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-s">python-polyCRACKER.py-generate_genome_signatures command line option</a>
</li>
      </ul></li>
      <li>
    -s, --sequences_dict &lt;sequences_dict&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-extract-sequences-s">python-polyCRACKER.py-extract_sequences command line option</a>
</li>
      </ul></li>
      <li>
    -s, --silhouette

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-s">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-s">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -s, --subclass_dir &lt;subclass_dir&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-compare-subclasses-s">python-polyCRACKER.py-compare_subclasses command line option</a>
</li>
      </ul></li>
      <li>
    -s, --subgenome_folder &lt;subgenome_folder&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-exchange-s">python-polyCRACKER.py-cluster_exchange command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-s">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -s, --submission_system &lt;submission_system&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-s">python-polyCRACKER.py-run_pipeline command line option</a>
</li>
      </ul></li>
      <li>
    -sbed, --progenitors_species_bed &lt;progenitors_species_bed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-sbed">python-polyCRACKER.py-final_stats command line option</a>
</li>
      </ul></li>
      <li>
    -scan, --hyperparameter_scan

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-scan">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-scan">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -sf, --scaffolds_file &lt;scaffolds_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-sf">python-polyCRACKER.py-clusterGraph command line option</a>
</li>
      </ul></li>
      <li>
    -sk, --skip_kcount

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-sk">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
      </ul></li>
      <li>
    -sl, --split_length &lt;split_length&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-sl">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-sl">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-sl">python-polyCRACKER.py-generate_unionbed command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-sl">python-polyCRACKER.py-plot_rules_chromosomes command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-sl">python-polyCRACKER.py-return_dash_data_structures command line option</a>
</li>
      </ul></li>
      <li>
    -sp, --input_species &lt;input_species&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-denovo-repeats-sp">python-polyCRACKER.py-find_denovo_repeats command line option</a>
</li>
      </ul></li>
      <li>
    -sp, --sparse_matrix_file &lt;sparse_matrix_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-sp">python-polyCRACKER.py-clusterGraph command line option</a>
</li>
      </ul></li>
      <li>
    -t, --kmer_type_path &lt;kmer_type_path&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-t">python-polyCRACKER.py-polyploid_diff_kmer_comparison command line option</a>
</li>
      </ul></li>
      <li>
    -t, --min_chunk_threshold &lt;min_chunk_threshold&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-t">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-t">python-polyCRACKER.py-splitFasta command line option</a>
</li>
      </ul></li>
      <li>
    -t, --technique &lt;technique&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-t">python-polyCRACKER.py-transform_plot command line option</a>
</li>
      </ul></li>
      <li>
    -t, --time &lt;time&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-t">python-polyCRACKER.py-run_pipeline command line option</a>
</li>
      </ul></li>
      <li>
    -t, --transform_algorithm &lt;transform_algorithm&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-t">python-polyCRACKER.py-plot_distance_matrix command line option</a>
</li>
      </ul></li>
      <li>
    -tf, --tfidf &lt;tfidf&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-tf">python-polyCRACKER.py-generate_Kmer_Matrix command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-tf">python-polyCRACKER.py-transform_plot command line option</a>
</li>
      </ul></li>
      <li>
    -txt, --txt_files &lt;txt_files&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-txt2fasta-txt">python-polyCRACKER.py-txt2fasta command line option</a>
</li>
      </ul></li>
      <li>
    -u, --unionbed &lt;unionbed&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-unionbed2matrix-u">python-polyCRACKER.py-unionbed2matrix command line option</a>
</li>
      </ul></li>
      <li>
    -u, --unionbed_file &lt;unionbed_file&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-unionbed-u">python-polyCRACKER.py-plot_unionbed command line option</a>
</li>
      </ul></li>
      <li>
    -u, --unionbed_threshold &lt;unionbed_threshold&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-u">python-polyCRACKER.py-subgenomeExtraction command line option</a>
</li>
      </ul></li>
      <li>
    -v, --validity

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-v">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-v">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
      </ul></li>
      <li>
    -w, --window_length &lt;window_length&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-get-density-w">python-polyCRACKER.py-get_density command line option</a>
</li>
      </ul></li>
      <li>
    -w, --work_dir &lt;work_dir&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-w">python-polyCRACKER.py-TE_Cluster_Analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-avg-distance-between-diff-kmers-w">python-polyCRACKER.py-avg_distance_between_diff_kmers command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-w">python-polyCRACKER.py-bio_hyp_class command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-w">python-polyCRACKER.py-categorize_repeats command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-color-trees-w">python-polyCRACKER.py-color_trees command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-w">python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genome-quality-assessment-w">python-polyCRACKER.py-dash_genome_quality_assessment command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-w">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-w">python-polyCRACKER.py-explore_kmers command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-rules-w">python-polyCRACKER.py-find_rules command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-rules2-w">python-polyCRACKER.py-find_rules2 command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-merge-split-kmer-clusters-w">python-polyCRACKER.py-merge_split_kmer_clusters command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-w">python-polyCRACKER.py-plot_rules command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-w">python-polyCRACKER.py-plot_rules_chromosomes command line option</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-w">python-polyCRACKER.py-repeat_subclass_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -w, --work_folder &lt;work_folder&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-w">python-polyCRACKER.py-generate_unionbed command line option</a>
</li>
      </ul></li>
      <li>
    -wnn, --weighted_nn &lt;weighted_nn&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-wnn">python-polyCRACKER.py-cluster command line option</a>
</li>
      </ul></li>
      <li>
    -wnn, --weights &lt;weights&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-wnn">python-polyCRACKER.py-diff_kmer_analysis command line option</a>
</li>
      </ul></li>
      <li>
    -y, --y_true_pickle &lt;y_true_pickle&gt;

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-y">python-polyCRACKER.py-find_best_cluster_parameters command line option</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="P">P</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    python-polyCRACKER.py command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-version">--version</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-anchor2bed command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-anchor2bed-a">-a, --anchor_file &lt;anchor_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-anchor2bed-q">-q, --qbed &lt;qbed&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-anchor2bed-s">-s, --sbed &lt;sbed&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-avg_distance_between_diff_kmers command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-avg-distance-between-diff-kmers-go">-go, --original_genome &lt;original_genome&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-avg-distance-between-diff-kmers-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-bed2scaffolds_pickle command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bed2scaffolds-pickle-i">-i, --bed_file &lt;bed_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bed2scaffolds-pickle-o">-o, --output_pickle &lt;output_pickle&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-bio_hyp_class command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-dk">-dk, --ratio_threshold &lt;ratio_threshold&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-f">-f, --folder_delimiter &lt;folder_delimiter&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-ft">-ft, --fast_tree</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-i">-i, --input_dict &lt;input_dict&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-kv">-kv, --default_kcount_value &lt;default_kcount_value&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-l">-l, --kmer_length &lt;kmer_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-max">-max, --max_val &lt;max_val&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-min">-min, --min_count &lt;min_count&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-sk">-sk, --skip_kcount</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-bio-hyp-class-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-blast2bed command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-b">-b, --bb &lt;bb&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-ex">-ex, --external_call</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-i">-i, --blast_file &lt;blast_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-l">-l, --low_memory &lt;low_memory&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-o">-o, --output_bed_file &lt;output_bed_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast2bed-om">-om, --output_merged_bed_file &lt;output_merged_bed_file&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-blast_kmers command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-k">-k, --kmer_fasta &lt;kmer_fasta&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-kl">-kl, --kmer_length &lt;kmer_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-o">-o, --output_file &lt;output_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-pm">-pm, --perfect_mode &lt;perfect_mode&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-blast-kmers-r">-r, --reference_genome &lt;reference_genome&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-categorize_repeats command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-dash">-dash, --output_dash</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-gff">-gff, --repeat_gff &lt;repeat_gff&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-l">-l, --kmer_length &lt;kmer_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-motif">-motif, --grab_motif</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-r">-r, --rules_kmers &lt;rules_kmers&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-rf">-rf, --repeat_fasta &lt;repeat_fasta&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-categorize-repeats-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-cluster command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-all">-all, --grab_all &lt;grab_all&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-c">-c, --cluster_method &lt;cluster_method&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-f">-f, --file &lt;file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-g">-g, --gamma &lt;gamma&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-m">-m, --metric &lt;metric&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-mst">-mst, --min_span_tree &lt;min_span_tree&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-nn">-nn, --n_neighbors &lt;n_neighbors&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-s">-s, --n_subgenomes &lt;n_subgenomes&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-wnn">-wnn, --weighted_nn &lt;weighted_nn&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-cluster_exchange command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-exchange-l">-l, --list_of_clusters &lt;list_of_clusters&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-cluster-exchange-s">-s, --subgenome_folder &lt;subgenome_folder&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-clusterGraph command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-b">-b, --bed_features_file &lt;bed_features_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-i">-i, --iteration &lt;iteration&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-l">-l, --layout &lt;layout&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-od">-od, --out_dir &lt;out_dir&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-p">-p, --positions_npy &lt;positions_npy&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-sf">-sf, --scaffolds_file &lt;scaffolds_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-clustergraph-sp">-sp, --sparse_matrix_file &lt;sparse_matrix_file&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-color_trees command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-color-trees-n">-n, --newick_in &lt;newick_in&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-color-trees-rd">-rd, --top_repeat_dict_pickle &lt;top_repeat_dict_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-color-trees-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-compare_subclasses command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-compare-subclasses-s">-s, --subclass_dir &lt;subclass_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-compareSubgenomes_progenitors_v_extracted command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-b1">-b1, --bed1 &lt;bed1&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-b2">-b2, --bed2 &lt;bed2&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-od">-od, --out_dir &lt;out_dir&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-s">-s, --scaffolds_file &lt;scaffolds_file&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-convert_mat2r command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-convert-mat2r-npz">-npz, --input_matrix &lt;input_matrix&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-convert_subgenome_output_to_pickle command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-id">-id, --input_dir &lt;input_dir&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-o">-o, --output_pickle &lt;output_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-s">-s, --scaffolds_pickle &lt;scaffolds_pickle&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-correct_kappa command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-correct-kappa-dict">-dict, --comparisons_dict &lt;comparisons_dict&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-correct-kappa-p0">-p0, --scaffolds_pickle &lt;scaffolds_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-correct-kappa-p1">-p1, --polycracker_pickle &lt;polycracker_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-correct-kappa-p2">-p2, --progenitors_pickle &lt;progenitors_pickle&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-dash_genetic_algorithm_hdbscan_test command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-csv">-csv, --kmer_count_matrix &lt;kmer_count_matrix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-d">-d, --n_dimensions &lt;n_dimensions&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-j">-j, --n_jobs &lt;n_jobs&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-k">-k, --kernel &lt;kernel&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-low">-low, --low_counts &lt;low_counts&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-m">-m, --metric &lt;metric&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-c">-min_c, --min_cluster_size &lt;min_cluster_size&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-n">-min_n, --min_number_clusters &lt;min_number_clusters&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-s">-min_s, --min_samples &lt;min_samples&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-nn">-nn, --n_neighbors &lt;n_neighbors&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-r">-r, --reduction_method &lt;reduction_method&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-s">-s, --silhouette</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-scan">-scan, --hyperparameter_scan</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-v">-v, --validity</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-dash_genome_quality_assessment command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genome-quality-assessment-csv">-csv, --kmer_rule_matrix &lt;kmer_rule_matrix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genome-quality-assessment-g">-g, --pre_scaffolded_genomes &lt;pre_scaffolded_genomes&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genome-quality-assessment-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-dash-genome-quality-assessment-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-diff_kmer_analysis command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-b">-b, --final_output_label_bed &lt;final_output_label_bed&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-dd">-dd, --diff_kmers_directory &lt;diff_kmers_directory&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-fp">-fp, --fasta_path &lt;fasta_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-go">-go, --original_genome &lt;original_genome&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-k">-k, --kernel &lt;kernel&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-kp">-kp, --kmers_pickle &lt;kmers_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-min">-min, --min_number_diff_kmers &lt;min_number_diff_kmers&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-n">-n, --n_chromosomes &lt;n_chromosomes&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-ndb">-ndb, --no_diff_blast</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-nj">-nj, --no_jaccard_clustering</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-nnk">-nnk, --n_neighbors_kmers &lt;n_neighbors_kmers&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-nns">-nns, --n_neighbors_scaffolds &lt;n_neighbors_scaffolds&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-np">-np, --no_run_phylogeny</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-npz">-npz, --sparse_kmer_matrix &lt;sparse_kmer_matrix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-ns">-ns, --no_sam2matrix</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-p">-p, --final_output_label_pickle &lt;final_output_label_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-r">-r, --repeat_masked_gff &lt;repeat_masked_gff&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-n-subgenomes">-s, --n_subgenomes &lt;n_subgenomes&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-s">-s, --scaffolds_pickle &lt;scaffolds_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-sl">-sl, --split_length &lt;split_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-diff-kmer-analysis-wnn">-wnn, --weights &lt;weights&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-differential_TE_histogram command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-differential-te-histogram-npy">-npy, --differential_repeat_subgenome_matrix &lt;differential_TE_subgenome_matrix&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-explore_kmers command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-csv">-csv, --kmer_count_matrix &lt;kmer_count_matrix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-d">-d, --n_dimensions &lt;n_dimensions&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-j">-j, --n_jobs &lt;n_jobs&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-k">-k, --kernel &lt;kernel&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-low">-low, --low_counts &lt;low_counts&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-m">-m, --metric &lt;metric&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-min-c">-min_c, --min_cluster_size &lt;min_cluster_size&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-min-n">-min_n, --min_number_clusters &lt;min_number_clusters&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-min-s">-min_s, --min_samples &lt;min_samples&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-nn">-nn, --n_neighbors &lt;n_neighbors&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-r">-r, --reduction_method &lt;reduction_method&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-s">-s, --silhouette</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-scan">-scan, --hyperparameter_scan</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-v">-v, --validity</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-explore-kmers-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-extract_sequences command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-extract-sequences-fi">-fi, --fasta_in &lt;fasta_in&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-extract-sequences-s">-s, --sequences_dict &lt;sequences_dict&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-final_stats command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-c">-c, --cluster_analysis</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-cbed">-cbed, --correspondence_bed &lt;correspondence_bed&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-cs">-cs, --cluster_scaffold &lt;cluster_scaffold&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-dict">-dict, --comparisons_dict &lt;comparisons_dict&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-n">-n, --note &lt;note&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-npy">-npy, --positions_npy &lt;positions_npy&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-op">-op, --original_progenitors &lt;original_progenitors&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-p">-p, --progenitors, -s, --species</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-pbed">-pbed, --polycracker_bed &lt;polycracker_bed&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-final-stats-sbed">-sbed, --progenitors_species_bed &lt;progenitors_species_bed&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-find_best_cluster_parameters command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-f">-f, --file &lt;file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-g">-g, --generations_number &lt;generations_number&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-gc">-gc, --gene_crossover_prob &lt;gene_crossover_prob&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-gm">-gm, --gene_mutation_prob &lt;gene_mutation_prob&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-p">-p, --population_size &lt;population_size&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-s">-s, --n_subgenomes &lt;n_subgenomes&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-best-cluster-parameters-y">-y, --y_true_pickle &lt;y_true_pickle&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-find_denovo_repeats command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-denovo-repeats-fi">-fi, --input_fasta &lt;input_fasta&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-denovo-repeats-od">-od, --out_dir &lt;out_dir&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-denovo-repeats-rd">-rd, --recover_dir &lt;recover_dir&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-denovo-repeats-sp">-sp, --input_species &lt;input_species&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-find_rules command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-rules-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-find_rules2 command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-rules2-n">-n, -n_rules &lt;n_rules&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-find-rules2-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-generate_genome_signatures command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-i">-i, --input_files &lt;input_files&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-kl">-kl, --kmer_length &lt;kmer_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-mf">-mf, --multi_fasta</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-o">-o, --output_file_prefix &lt;output_file_prefix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-genome-signatures-s">-s, --scaled &lt;scaled&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-generate_karyotype command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-karyotype-shiny">--shiny</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-karyotype-go">-go, --original_genome &lt;original_genome&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-generate_Kmer_Matrix command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-c">-c, --chunk_size &lt;chunk_size&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-d">-d, --kmercount_path &lt;kmercount_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-f">-f, --fasta_path &lt;fasta_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-g">-g, --genome &lt;genome&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-go">-go, --original_genome &lt;original_genome&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-gs">-gs, --genome_split_name &lt;genome_split_name&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-i">-i, --input_bed_file &lt;input_bed_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-im">-im, --input_merged_bed_file &lt;input_merged_bed_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-k">-k, --output_kmer_file &lt;output_kmer_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-l">-l, --low_memory &lt;low_memory&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-mc">-mc, --min_chunk_size &lt;min_chunk_size&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-npz">-npz, --output_sparse_matrix &lt;output_sparse_matrix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-p">-p, --prefilter &lt;prefilter&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-r">-r, --remove_non_chunk &lt;remove_non_chunk&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-t">-t, --min_chunk_threshold &lt;min_chunk_threshold&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-kmer-matrix-tf">-tf, --tfidf &lt;tfidf&gt;</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    python-polyCRACKER.py-generate_out_bed command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-out-bed-fai">-fai, --fai_file &lt;fai_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-out-bed-fo">-fo, --output_fname &lt;output_fname&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-out-bed-id">-id, --input_dir &lt;input_dir&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-out-bed-o">-o, --original &lt;original&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-generate_unionbed command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-b3">-b3, --bed3_directory &lt;bed3_directory&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-fai">-fai, --fai_file &lt;fai_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-o">-o, --original &lt;original&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-sl">-sl, --split_length &lt;split_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-generate-unionbed-w">-w, --work_folder &lt;work_folder&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-genomic_label_propagation command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-c">-c, --input_labels &lt;input_labels&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-npy">-npy, --transformed_data &lt;transformed_data&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-o">-o, --output_pickle &lt;output_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-r">-r, --radius &lt;radius&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-genomic-label-propagation-s">-s, --scaffolds_pickle &lt;scaffolds_pickle&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-get_density command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-get-density-fai">-fai, --fai_file &lt;fai_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-get-density-fout">-fout, --outputfname &lt;outputfname&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-get-density-gff">-gff, --gff_file &lt;gff_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-get-density-w">-w, --window_length &lt;window_length&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-hipmer_output_to_kcount command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-hipmer-output-to-kcount-d">-d, --run_on_dir</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-hipmer-output-to-kcount-i">-i, --hipmer_input &lt;hipmer_input&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-hipmer-output-to-kcount-o">-o, --kcount_output &lt;kcount_output&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-kcount_hist command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-log">--log</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-k">-k, --kcount_directory &lt;kcount_directory&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-kl">-kl, --kmer_lengths &lt;kmer_lengths&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-l">-l, --low_count &lt;low_count&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-o">-o, --out_file &lt;out_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-r">-r, --kcount_range &lt;kcount_range&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-s">-s, --sample_speed &lt;sample_speed&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-kcount_hist_old command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-old-log">--log</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-old-k">-k, --kcount_file &lt;kcount_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kcount-hist-old-o">-o, --out_file &lt;out_file&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-kmer2Fasta command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-hb">-hb, --high_bool &lt;high_bool&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-hc">-hc, --kmer_high_count &lt;kmer_high_count&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-k">-k, --kmercount_path &lt;kmercount_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-lc">-lc, --kmer_low_count &lt;kmer_low_count&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-kmer2fasta-s">-s, --sampling_sensitivity &lt;sampling_sensitivity&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-link2color command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-link2color-csv">-csv, --link_csv &lt;link_csv&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-merge_split_kmer_clusters command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-merge-split-kmer-clusters-c">-c, --estimated_clusters &lt;estimated_clusters&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-merge-split-kmer-clusters-d">-d, --n_dimensions &lt;n_dimensions&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-merge-split-kmer-clusters-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-multicol2multifiles command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-multicol2multifiles-csv">-csv, --multi_column_csv &lt;multi_column_csv&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-out_bed_to_circos_csv command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-out-bed-to-circos-csv-b">-b, --output_bed &lt;output_bed&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-out-bed-to-circos-csv-fai">-fai, --fai_file &lt;fai_file&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-plot_distance_matrix command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-csv">-csv, --dist_mat &lt;dist_mat&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-f">-f, --feature_space &lt;feature_space&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-j">-j, --n_jobs &lt;n_jobs&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-k">-k, --kmeans_clusters &lt;kmeans_clusters&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-o">-o, --output_file_prefix &lt;output_file_prefix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-distance-matrix-t">-t, --transform_algorithm &lt;transform_algorithm&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-plot_rules command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-csv">-csv, --rule_csv &lt;rule_csv&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-plot_rules_chromosomes command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-csv">-csv, --kmer_rule_matrix &lt;kmer_rule_matrix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-go">-go, --original_genome &lt;original_genome&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-sl">-sl, --split_length &lt;split_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-rules-chromosomes-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-plot_unionbed command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-unionbed-n">-n, --number_chromosomes &lt;number_chromosomes&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-unionbed-od">-od, --out_folder &lt;out_folder&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plot-unionbed-u">-u, --unionbed_file &lt;unionbed_file&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-plotPositions command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-ax">-ax, --axes_off</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-c">-c, --colors_pickle &lt;colors_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-i">-i, --iterations &lt;iterations&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-l">-l, --layout &lt;layout&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-npy">-npy, --positions_npy &lt;positions_npy&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-npz">-npz, --graph_file &lt;graph_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-o">-o, --output_fname &lt;output_fname&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-p">-p, --labels_pickle &lt;labels_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-plotpositions-s">-s, --graph_sampling_speed &lt;graph_sampling_speed&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-polyploid_diff_kmer_comparison command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-l">-l, --kmer_list_path &lt;kmer_list_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-p">-p, --analysis_path &lt;analysis_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-t">-t, --kmer_type_path &lt;kmer_type_path&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-progenitorMapping command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-fout">-fout, --outputfname &lt;outputfname&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-gff">-gff, --gff_file &lt;gff_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-go">-go, --original_genome &lt;original_genome&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-i">-i, --input_fasta &lt;input_fasta&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-kl">-kl, --kmer_length &lt;kmer_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-o">-o, --original &lt;original&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-p">-p, --progenitor_fasta_folder &lt;progenitor_fasta_folder&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-progenitormapping-pf">-pf, --progenitor_fastas &lt;progenitor_fastas&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-repeat_subclass_analysis command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-bo">-bo, --original_subgenomes_bed &lt;original_subgenomes_bed&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-bt">-bt, --bootstrap</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-go">-go, --original_fasta &lt;original_fasta&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-r">-r, --all_family_subclass_pickle &lt;all_family_subclass_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-rd">-rd, --top_repeat_dict_pickle &lt;top_repeat_dict_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-re">-re, --all_repeat_elements &lt;all_repeat_elements&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-rf">-rf, --repeat_fasta &lt;repeat_fasta&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-rp">-rp, --top_repeat_list_pickle &lt;top_repeat_list_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-rs">-rs, --class_subclass_pickle &lt;class_subclass_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-repeat-subclass-analysis-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-return_dash_data_structures command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-csv">-csv, --kmer_rule_matrix &lt;kmer_rule_matrix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-kl">-kl, --kmer_pca_labels &lt;kmer_pca_labels&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-kr">-kr, --kmer_rules_pickle &lt;kmer_rules_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-npy">-npy, --kmer_pca_data &lt;kmer_pca_data&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-o">-o, --output_dir &lt;output_dir&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-r">-r, --reference_genome &lt;reference_genome&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-return-dash-data-structures-sl">-sl, --split_length &lt;split_length&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-run_iqtree command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-iqtree-b">-b, --bootstrap &lt;bootstrap&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-iqtree-f">-f, --fasta_in &lt;fasta_in&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-iqtree-m">-m, --model &lt;model&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-iqtree-nt">-nt, --n_threads &lt;n_threads&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-run_pipeline command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-j">-J, --job_name &lt;job_name&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-n">-N, --nodes &lt;nodes&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-a">-a, --account_name &lt;account_name&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-b">-b, --batch_script &lt;batch_script&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-env">-env, --environment &lt;environment&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-m">-m, --memory &lt;memory&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-no-options">-n, --no_options</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-s">-s, --submission_system &lt;submission_system&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-run-pipeline-t">-t, --time &lt;time&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-scaffolds2colors_specified command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-scaffolds2colors-specified-d">-d, --labels_dict &lt;labels_dict&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-scaffolds2colors-specified-o">-o, --output_pickle &lt;output_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-scaffolds2colors-specified-s">-s, --scaffolds_pickle &lt;scaffolds_pickle&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-send_repeats command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-send-repeats-rf">-rf, --subclass_repeat_fasta &lt;subclass_repeat_fasta&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-shiny2omic command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-shiny2omic-csv">-csv, --input_csv &lt;input_csv&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-species_comparison_scaffold2colors command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-species-comparison-scaffold2colors-o">-o, --output_pickle &lt;output_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-species-comparison-scaffold2colors-s">-s, --scaffolds_pickle &lt;scaffolds_pickle&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-spectral_embed_plot command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-spectral-embed-plot-npy">-npy, --positions &lt;positions&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-splitFasta command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-c">-c, --chunk_size &lt;chunk_size&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-d">-d, --fasta_path &lt;fasta_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-i">-i, --fasta_file &lt;fasta_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-l">-l, --low_memory &lt;low_memory&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-m">-m, --min_chunk_size &lt;min_chunk_size&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-p">-p, --prefilter &lt;prefilter&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-r">-r, --remove_non_chunk &lt;remove_non_chunk&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-splitfasta-t">-t, --min_chunk_threshold &lt;min_chunk_threshold&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-subgenomeExtraction command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-b">-b, --bootstrap &lt;bootstrap&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-bb">-bb, --bb &lt;bb&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-dk">-dk, --diff_kmer_threshold &lt;diff_kmer_threshold&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-ds">-ds, --diff_sample_rate &lt;diff_sample_rate&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-g">-g, --genome_name &lt;genome_name&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-go">-go, --original_genome &lt;original_genome&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-i">-i, --iteration &lt;iteration&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-kl">-kl, --kmer_low_count &lt;kmer_low_count&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-kv">-kv, --default_kmercount_value &lt;default_kmercount_value&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-l">-l, --kmer_length &lt;kmer_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-o">-o, --original &lt;original&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-os">-os, --original_subgenome_path &lt;original_subgenome_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-p">-p, --fasta_path &lt;fasta_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-r">-r, --run_final &lt;run_final&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-s">-s, --subgenome_folder &lt;subgenome_folder&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-subgenomeextraction-u">-u, --unionbed_threshold &lt;unionbed_threshold&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-TE_Cluster_Analysis command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-bo">-bo, --original_subgenomes_bed &lt;original_subgenomes_bed&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-d">-d, --n_dimensions &lt;n_dimensions&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-dt">-dt, --diff_repeat_threshold &lt;diff_repeat_threshold&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-dv">-dv, --default_repeat_val &lt;default_repeat_val&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-fai">-fai, --fai_file &lt;fai_file&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-k">-k, --kernel &lt;kernel&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-n">-n, --n_chromosomes &lt;n_chromosomes&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-nt">-nt, --n_top_repeats &lt;n_top_repeats&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-rg">-rg, --repeat_gff &lt;repeat_gff&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-rp">-rp, --repeat_pickle &lt;repeat_pickle&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-s">-s, --n_clusters &lt;n_clusters&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-sl">-sl, --split_length &lt;split_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-te-cluster-analysis-w">-w, --work_dir &lt;work_dir&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-transform_plot command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-d">-d, --n_dimensions &lt;n_dimensions&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-m">-m, --metric &lt;metric&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-s">-s, --n_subgenomes &lt;n_subgenomes&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-t">-t, --technique &lt;technique&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-transform-plot-tf">-tf, --tfidf &lt;tfidf&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-txt2fasta command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-txt2fasta-rf">-rf, --reference_fasta &lt;reference_fasta&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-txt2fasta-txt">-txt, --txt_files &lt;txt_files&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-unionbed2matrix command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-unionbed2matrix-o">-o, --output_matrix &lt;output_matrix&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-unionbed2matrix-u">-u, --unionbed &lt;unionbed&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-update_nextflow_config command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-update-nextflow-config-n">-n, --nextflow_config &lt;nextflow_config&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-update-nextflow-config-p">-p, --polycracker_config &lt;polycracker_config&gt;</a>
</li>
      </ul></li>
      <li>
    python-polyCRACKER.py-writeKmerCount command line option

      <ul>
        <li><a href="index.html#cmdoption-python-polycracker-py-writekmercount-d">-d, --fasta_path &lt;fasta_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-writekmercount-k">-k, --kmercount_path &lt;kmercount_path&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-writekmercount-l">-l, --kmer_length &lt;kmer_length&gt;</a>
</li>
        <li><a href="index.html#cmdoption-python-polycracker-py-writekmercount-m">-m, --blast_mem &lt;blast_mem&gt;</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>



           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Joshua Levy, Sean Gordon.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'v1.1.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>