

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Welcome to polyCRACKER’s documentation! &mdash; polyCRACKER v1.1.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="polyCRACKER v1.1.1 documentation" href="#"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="#" class="icon icon-home"> polyCRACKER
          

          
          </a>

          
            
            
              <div class="version">
                v1.1.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Welcome to polyCRACKER’s documentation!</a><ul>
<li><a class="reference internal" href="#python-polycracker-py">python polyCRACKER.py</a><ul>
<li><a class="reference internal" href="#python-polycracker-py-te-cluster-analysis">TE_Cluster_Analysis</a></li>
<li><a class="reference internal" href="#python-polycracker-py-anchor2bed">anchor2bed</a></li>
<li><a class="reference internal" href="#python-polycracker-py-avg-distance-between-diff-kmers">avg_distance_between_diff_kmers</a></li>
<li><a class="reference internal" href="#python-polycracker-py-bed2scaffolds-pickle">bed2scaffolds_pickle</a></li>
<li><a class="reference internal" href="#python-polycracker-py-bio-hyp-class">bio_hyp_class</a></li>
<li><a class="reference internal" href="#python-polycracker-py-blast2bed">blast2bed</a></li>
<li><a class="reference internal" href="#python-polycracker-py-blast-kmers">blast_kmers</a></li>
<li><a class="reference internal" href="#python-polycracker-py-build-pairwise-align-similarity-structure">build_pairwise_align_similarity_structure</a></li>
<li><a class="reference internal" href="#python-polycracker-py-categorize-repeats">categorize_repeats</a></li>
<li><a class="reference internal" href="#python-polycracker-py-cluster">cluster</a></li>
<li><a class="reference internal" href="#python-polycracker-py-clustergraph">clusterGraph</a></li>
<li><a class="reference internal" href="#python-polycracker-py-cluster-exchange">cluster_exchange</a></li>
<li><a class="reference internal" href="#python-polycracker-py-color-trees">color_trees</a></li>
<li><a class="reference internal" href="#python-polycracker-py-comparesubgenomes-progenitors-v-extracted">compareSubgenomes_progenitors_v_extracted</a></li>
<li><a class="reference internal" href="#python-polycracker-py-compare-subclasses">compare_subclasses</a></li>
<li><a class="reference internal" href="#python-polycracker-py-convert-mat2r">convert_mat2r</a></li>
<li><a class="reference internal" href="#python-polycracker-py-convert-subgenome-output-to-pickle">convert_subgenome_output_to_pickle</a></li>
<li><a class="reference internal" href="#python-polycracker-py-correct-kappa">correct_kappa</a></li>
<li><a class="reference internal" href="#python-polycracker-py-dash-genetic-algorithm-hdbscan-test">dash_genetic_algorithm_hdbscan_test</a></li>
<li><a class="reference internal" href="#python-polycracker-py-dash-genome-quality-assessment">dash_genome_quality_assessment</a></li>
<li><a class="reference internal" href="#python-polycracker-py-diff-kmer-analysis">diff_kmer_analysis</a></li>
<li><a class="reference internal" href="#python-polycracker-py-differential-te-histogram">differential_TE_histogram</a></li>
<li><a class="reference internal" href="#python-polycracker-py-explore-kmers">explore_kmers</a></li>
<li><a class="reference internal" href="#python-polycracker-py-extract-sequences">extract_sequences</a></li>
<li><a class="reference internal" href="#python-polycracker-py-final-stats">final_stats</a></li>
<li><a class="reference internal" href="#python-polycracker-py-find-best-cluster-parameters">find_best_cluster_parameters</a></li>
<li><a class="reference internal" href="#python-polycracker-py-find-denovo-repeats">find_denovo_repeats</a></li>
<li><a class="reference internal" href="#python-polycracker-py-find-rules">find_rules</a></li>
<li><a class="reference internal" href="#python-polycracker-py-find-rules2">find_rules2</a></li>
<li><a class="reference internal" href="#python-polycracker-py-generate-kmer-matrix">generate_Kmer_Matrix</a></li>
<li><a class="reference internal" href="#python-polycracker-py-generate-genome-signatures">generate_genome_signatures</a></li>
<li><a class="reference internal" href="#python-polycracker-py-generate-karyotype">generate_karyotype</a></li>
<li><a class="reference internal" href="#python-polycracker-py-generate-out-bed">generate_out_bed</a></li>
<li><a class="reference internal" href="#python-polycracker-py-generate-unionbed">generate_unionbed</a></li>
<li><a class="reference internal" href="#python-polycracker-py-genomic-label-propagation">genomic_label_propagation</a></li>
<li><a class="reference internal" href="#python-polycracker-py-get-density">get_density</a></li>
<li><a class="reference internal" href="#python-polycracker-py-hipmer-output-to-kcount">hipmer_output_to_kcount</a></li>
<li><a class="reference internal" href="#python-polycracker-py-kcount-hist">kcount_hist</a></li>
<li><a class="reference internal" href="#python-polycracker-py-kcount-hist-old">kcount_hist_old</a></li>
<li><a class="reference internal" href="#python-polycracker-py-kmer2fasta">kmer2Fasta</a></li>
<li><a class="reference internal" href="#python-polycracker-py-kmerratio2scaffasta">kmerratio2scaffasta</a></li>
<li><a class="reference internal" href="#python-polycracker-py-link2color">link2color</a></li>
<li><a class="reference internal" href="#python-polycracker-py-mash-test">mash_test</a></li>
<li><a class="reference internal" href="#python-polycracker-py-merge-split-kmer-clusters">merge_split_kmer_clusters</a></li>
<li><a class="reference internal" href="#python-polycracker-py-multicol2multifiles">multicol2multifiles</a></li>
<li><a class="reference internal" href="#python-polycracker-py-out-bed-to-circos-csv">out_bed_to_circos_csv</a></li>
<li><a class="reference internal" href="#python-polycracker-py-plotpositions">plotPositions</a></li>
<li><a class="reference internal" href="#python-polycracker-py-plot-distance-matrix">plot_distance_matrix</a></li>
<li><a class="reference internal" href="#python-polycracker-py-plot-rules">plot_rules</a></li>
<li><a class="reference internal" href="#python-polycracker-py-plot-rules-chromosomes">plot_rules_chromosomes</a></li>
<li><a class="reference internal" href="#python-polycracker-py-plot-unionbed">plot_unionbed</a></li>
<li><a class="reference internal" href="#python-polycracker-py-polyploid-diff-kmer-comparison">polyploid_diff_kmer_comparison</a></li>
<li><a class="reference internal" href="#python-polycracker-py-progenitormapping">progenitorMapping</a></li>
<li><a class="reference internal" href="#python-polycracker-py-repeat-subclass-analysis">repeat_subclass_analysis</a></li>
<li><a class="reference internal" href="#python-polycracker-py-reset-cluster">reset_cluster</a></li>
<li><a class="reference internal" href="#python-polycracker-py-reset-transform">reset_transform</a></li>
<li><a class="reference internal" href="#python-polycracker-py-return-dash-data-structures">return_dash_data_structures</a></li>
<li><a class="reference internal" href="#python-polycracker-py-run-iqtree">run_iqtree</a></li>
<li><a class="reference internal" href="#python-polycracker-py-run-pipeline">run_pipeline</a></li>
<li><a class="reference internal" href="#python-polycracker-py-run-tests">run_tests</a></li>
<li><a class="reference internal" href="#python-polycracker-py-scaffolds2colors-specified">scaffolds2colors_specified</a></li>
<li><a class="reference internal" href="#python-polycracker-py-send-repeats">send_repeats</a></li>
<li><a class="reference internal" href="#python-polycracker-py-shiny2omic">shiny2omic</a></li>
<li><a class="reference internal" href="#python-polycracker-py-species-comparison-scaffold2colors">species_comparison_scaffold2colors</a></li>
<li><a class="reference internal" href="#python-polycracker-py-spectral-embed-plot">spectral_embed_plot</a></li>
<li><a class="reference internal" href="#python-polycracker-py-splitfasta">splitFasta</a></li>
<li><a class="reference internal" href="#python-polycracker-py-subgenomeextraction">subgenomeExtraction</a></li>
<li><a class="reference internal" href="#python-polycracker-py-transform-plot">transform_plot</a></li>
<li><a class="reference internal" href="#python-polycracker-py-txt2fasta">txt2fasta</a></li>
<li><a class="reference internal" href="#python-polycracker-py-unionbed2matrix">unionbed2matrix</a></li>
<li><a class="reference internal" href="#python-polycracker-py-update-nextflow-config">update_nextflow_config</a></li>
<li><a class="reference internal" href="#python-polycracker-py-writekmercount">writeKmerCount</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">polyCRACKER</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>Welcome to polyCRACKER’s documentation!</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="welcome-to-polycracker-s-documentation">
<h1>Welcome to polyCRACKER’s documentation!<a class="headerlink" href="#welcome-to-polycracker-s-documentation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
</div>
<div class="section" id="python-polycracker-py">
<h2>python polyCRACKER.py<a class="headerlink" href="#python-polycracker-py" title="Permalink to this headline">¶</a></h2>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-version">
<code class="descname">--version</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-version" title="Permalink to this definition">¶</a></dt>
<dd><p>Show the version and exit.</p>
</dd></dl>

<div class="section" id="python-polycracker-py-te-cluster-analysis">
<h3>TE_Cluster_Analysis<a class="headerlink" href="#python-polycracker-py-te-cluster-analysis" title="Permalink to this headline">¶</a></h3>
<p>Build clustering matrix (repeat counts vs scaffolds) from repeats and conduct analyses on repeats instead of kmers.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py TE_Cluster_Analysis <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for TE clustering is done.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-rg">
<code class="descname">-rg</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--repeat_gff</code><code class="descclassname"> &lt;repeat_gff&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-rg" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to repeat TE gff file, output from repeat masker pipeline.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-rp">
<code class="descname">-rp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--repeat_pickle</code><code class="descclassname"> &lt;repeat_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-rp" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to repeat TE pickle file, containing TE names, which will be generated in this pipeline.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-sl">
<code class="descname">-sl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--split_length</code><code class="descclassname"> &lt;split_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-sl" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of intervals in bedgraph files.  [default: 75000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_chromosomes</code><code class="descclassname"> &lt;n_chromosomes&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Output unionbed plots of x largest chromosomes.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_clusters</code><code class="descclassname"> &lt;n_clusters&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of clusters for clustering TEs.  [default: 2]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_dimensions</code><code class="descclassname"> &lt;n_dimensions&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of dimensions to reduce to.  [default: 3]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kernel</code><code class="descclassname"> &lt;kernel&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Kernel for KPCA. Cosine is particularly effective.  [default: cosine]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-nt">
<code class="descname">-nt</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_top_repeats</code><code class="descclassname"> &lt;n_top_repeats&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-nt" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of top TEs in each subgenome to analyze.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-fai">
<code class="descname">-fai</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fai_file</code><code class="descclassname"> &lt;fai_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-fai" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original, nonchunked fai file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-dv">
<code class="descname">-dv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--default_repeat_val</code><code class="descclassname"> &lt;default_repeat_val&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-dv" title="Permalink to this definition">¶</a></dt>
<dd><p>Default value for number of times TE appears in subgenome if value is 0.  [default: 2]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-dt">
<code class="descname">-dt</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--diff_repeat_threshold</code><code class="descclassname"> &lt;diff_repeat_threshold&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-dt" title="Permalink to this definition">¶</a></dt>
<dd><p>Threshold specifying that if one TE is differential in one subgenome versus others label that TE by that subgenome.  [default: 20]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-te-cluster-analysis-bo">
<code class="descname">-bo</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_subgenomes_bed</code><code class="descclassname"> &lt;original_subgenomes_bed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-te-cluster-analysis-bo" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original bed file containing subgenome labelling. Will use this to label TEs by subgenome instead of clustering. Leave blank if deciding labels via clustering.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-anchor2bed">
<h3>anchor2bed<a class="headerlink" href="#python-polycracker-py-anchor2bed" title="Permalink to this headline">¶</a></h3>
<p>Convert syntenic blocks of genes to bed coordinates between the two genomes being compared.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py anchor2bed <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-anchor2bed-a">
<code class="descname">-a</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--anchor_file</code><code class="descclassname"> &lt;anchor_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-anchor2bed-a" title="Permalink to this definition">¶</a></dt>
<dd><p>Lifted anchor file generated from basic synteny run using jcvi tools.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-anchor2bed-q">
<code class="descname">-q</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--qbed</code><code class="descclassname"> &lt;qbed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-anchor2bed-q" title="Permalink to this definition">¶</a></dt>
<dd><p>First bed file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-anchor2bed-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--sbed</code><code class="descclassname"> &lt;sbed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-anchor2bed-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Second bed file.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-avg-distance-between-diff-kmers">
<h3>avg_distance_between_diff_kmers<a class="headerlink" href="#python-polycracker-py-avg-distance-between-diff-kmers" title="Permalink to this headline">¶</a></h3>
<p>Report the average distance between iterations of unique differential kmers over any scaffold.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py avg_distance_between_diff_kmers <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-avg-distance-between-diff-kmers-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-avg-distance-between-diff-kmers-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for differential kmer analysis is done.  [default: ./diff_kmer_distances/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-avg-distance-between-diff-kmers-go">
<code class="descname">-go</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_genome</code><code class="descclassname"> &lt;original_genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-avg-distance-between-diff-kmers-go" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original polyploid fasta, prechunked.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-bed2scaffolds-pickle">
<h3>bed2scaffolds_pickle<a class="headerlink" href="#python-polycracker-py-bed2scaffolds-pickle" title="Permalink to this headline">¶</a></h3>
<p>Convert correspondence bed file, containing complete scaffold information into a list of scaffolds file.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py bed2scaffolds_pickle <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-bed2scaffolds-pickle-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bed_file</code><code class="descclassname"> &lt;bed_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bed2scaffolds-pickle-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Input bed file, original coordinate system.  [default: correspondence.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bed2scaffolds-pickle-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_pickle</code><code class="descclassname"> &lt;output_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bed2scaffolds-pickle-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file containing the new scaffolds.  [default: scaffolds_new.p]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-bio-hyp-class">
<h3>bio_hyp_class<a class="headerlink" href="#python-polycracker-py-bio-hyp-class" title="Permalink to this headline">¶</a></h3>
<p>Generate count matrix of kmers versus genomes. Optionally can transpose this matrix and find kmer guide tree / approximated phylogeny based on differential kmer counts. Changing settings to include all kmers can generate a more accurate guide tree.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py bio_hyp_class <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_dict</code><code class="descclassname"> &lt;input_dict&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-i" title="Permalink to this definition">¶</a></dt>
<dd><p>MANDATORY: Comma delimited mapping of all subgenome names to their subgenome fasta files.  [default: subgenome_1:./fasta_files/subgenome_1.fa,SpeciesA:./SpeciesA.fasta]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-f">
<code class="descname">-f</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--folder_delimiter</code><code class="descclassname"> &lt;folder_delimiter&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-f" title="Permalink to this definition">¶</a></dt>
<dd><p>OPTIONA: Comma delimited list of folder containing genomes, delimiters from which to construct short names, and number of characters to use for short name. Save fastas as AAAA_xxx_v0.fa format. Supercedes -i option, unless default value not modified.  [default: ./folder/,_,4]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-kv">
<code class="descname">-kv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--default_kcount_value</code><code class="descclassname"> &lt;default_kcount_value&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-kv" title="Permalink to this definition">¶</a></dt>
<dd><p>If kmer hits in subgenome is 0, this value will replace it.  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-dk">
<code class="descname">-dk</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--ratio_threshold</code><code class="descclassname"> &lt;ratio_threshold&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-dk" title="Permalink to this definition">¶</a></dt>
<dd><p>Value indicates that if the counts of a particular kmer in a subgenome divided by any of the counts in other subgenomes are greater than this value, then the kmer is differential.  [default: 20]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for biological hypothesis testing is done.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_length</code><code class="descclassname"> &lt;kmer_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of kmers to find.  [default: 23]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-ft">
<code class="descname">-ft</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fast_tree</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-ft" title="Permalink to this definition">¶</a></dt>
<dd><p>Make quick estimated phylogenetic tree out of kmer matrix.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-min">
<code class="descname">-min</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_count</code><code class="descclassname"> &lt;min_count&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-min" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum count for kmer counting.  [default: 3]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-sk">
<code class="descname">-sk</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--skip_kcount</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-sk" title="Permalink to this definition">¶</a></dt>
<dd><p>Skip generation of kcount files. kcount files must be named by the following convention: subgenome_1 used in input dict -&gt; subgenome_1.kcount</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-bio-hyp-class-max">
<code class="descname">-max</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--max_val</code><code class="descclassname"> &lt;max_val&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-bio-hyp-class-max" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum count for maximum count of kmer between subgenomes.  [default: 100]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-blast2bed">
<h3>blast2bed<a class="headerlink" href="#python-polycracker-py-blast2bed" title="Permalink to this headline">¶</a></h3>
<p>Converts the blast results from blast or bbtools into a bed file</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py blast2bed <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast2bed-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_file</code><code class="descclassname"> &lt;blast_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast2bed-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file after blasting kmers to genome. Must be input into this command.  [default: ./blast_files/results.sam]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast2bed-b">
<code class="descname">-b</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bb</code><code class="descclassname"> &lt;bb&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast2bed-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether bbtools were used in generating blasted sam file.  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast2bed-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--low_memory</code><code class="descclassname"> &lt;low_memory&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast2bed-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Do not merge the bed file. Takes longer to process but uses lower memory.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast2bed-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_bed_file</code><code class="descclassname"> &lt;output_bed_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast2bed-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output bed file, not merged.  [default: blasted.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast2bed-om">
<code class="descname">-om</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_merged_bed_file</code><code class="descclassname"> &lt;output_merged_bed_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast2bed-om" title="Permalink to this definition">¶</a></dt>
<dd><p>Output bed file, merged.  [default: blasted_merged.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast2bed-ex">
<code class="descname">-ex</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--external_call</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast2bed-ex" title="Permalink to this definition">¶</a></dt>
<dd><p>External call from non-polyCRACKER pipeline.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-blast-kmers">
<h3>blast_kmers<a class="headerlink" href="#python-polycracker-py-blast-kmers" title="Permalink to this headline">¶</a></h3>
<p>Maps kmers fasta file to a reference genome.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py blast_kmers <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast-kmers-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast-kmers-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast-kmers-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--reference_genome</code><code class="descclassname"> &lt;reference_genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast-kmers-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Genome to blast against.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast-kmers-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_fasta</code><code class="descclassname"> &lt;kmer_fasta&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast-kmers-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmer fasta converted from kmer count file</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast-kmers-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_file</code><code class="descclassname"> &lt;output_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast-kmers-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output sam file.  [default: results.sam]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast-kmers-kl">
<code class="descname">-kl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_length</code><code class="descclassname"> &lt;kmer_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast-kmers-kl" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmer length for mapping.  [default: 13]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-blast-kmers-pm">
<code class="descname">-pm</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--perfect_mode</code><code class="descclassname"> &lt;perfect_mode&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-blast-kmers-pm" title="Permalink to this definition">¶</a></dt>
<dd><p>Perfect mode.  [default: 1]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-build-pairwise-align-similarity-structure">
<h3>build_pairwise_align_similarity_structure<a class="headerlink" href="#python-polycracker-py-build-pairwise-align-similarity-structure" title="Permalink to this headline">¶</a></h3>
<p>Take consensus repeats, generate graph structure from sequence similarity, and cluster. Can be useful to find homologous repeats from different genomes.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py build_pairwise_align_similarity_structure 
    <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="python-polycracker-py-categorize-repeats">
<h3>categorize_repeats<a class="headerlink" href="#python-polycracker-py-categorize-repeats" title="Permalink to this headline">¶</a></h3>
<p>Take outputs from denovo repeat finding/modeling and intersect kmers of particular conservation patterns with these repeats,
either all iterations of a consensus repeat or consensus repeats themselves, and crosstabulate rule distribution with repeats/subclass.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py categorize_repeats <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-categorize-repeats-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-categorize-repeats-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for merging of clusters is done.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-categorize-repeats-rf">
<code class="descname">-rf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--repeat_fasta</code><code class="descclassname"> &lt;repeat_fasta&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-categorize-repeats-rf" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to repeat fasta file, output from repeat masker pipeline.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-categorize-repeats-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-categorize-repeats-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-categorize-repeats-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_length</code><code class="descclassname"> &lt;kmer_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-categorize-repeats-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of kmers to find.  [default: 23]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-categorize-repeats-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--rules_kmers</code><code class="descclassname"> &lt;rules_kmers&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-categorize-repeats-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmer count matrix with appended rule labels.  [default: ./kmer_master_matrix_rules.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-categorize-repeats-gff">
<code class="descname">-gff</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--repeat_gff</code><code class="descclassname"> &lt;repeat_gff&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-categorize-repeats-gff" title="Permalink to this definition">¶</a></dt>
<dd><p>If repeat gff is submitted here, repeat fasta must be original genome, and non-consensus repeat sequences will be considered.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-categorize-repeats-motif">
<code class="descname">-motif</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--grab_motif</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-categorize-repeats-motif" title="Permalink to this definition">¶</a></dt>
<dd><p>Grab unique consensus repeat identifiers.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-categorize-repeats-dash">
<code class="descname">-dash</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dash</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-categorize-repeats-dash" title="Permalink to this definition">¶</a></dt>
<dd><p>Output csv file for dash.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-cluster">
<h3>cluster<a class="headerlink" href="#python-polycracker-py-cluster" title="Permalink to this headline">¶</a></h3>
<p>Perform clustering on the dimensionality reduced data, though various methods. Takes in npy position file and outputs clustered regions and plot.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py cluster <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-f">
<code class="descname">-f</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--file</code><code class="descclassname"> &lt;file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-f" title="Permalink to this definition">¶</a></dt>
<dd><p>Numpy .npy file containing positions after dimensionality reduction</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cluster_method</code><code class="descclassname"> &lt;cluster_method&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Clustering method to use.  [default: SpectralClustering]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_subgenomes</code><code class="descclassname"> &lt;n_subgenomes&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of subgenomes  [default: 2]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--metric</code><code class="descclassname"> &lt;metric&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Distance metric used to compute affinity matrix, used to find nearest neighbors graph for spectral clustering.  [default: cosine]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-nn">
<code class="descname">-nn</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_neighbors</code><code class="descclassname"> &lt;n_neighbors&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-nn" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of nearest neighbors in generation of nearest neighbor graph.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-wnn">
<code class="descname">-wnn</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--weighted_nn</code><code class="descclassname"> &lt;weighted_nn&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-wnn" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to weight spectral graph for spectral clustering.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-all">
<code class="descname">-all</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--grab_all</code><code class="descclassname"> &lt;grab_all&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-all" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to grab all clusters, and number of clusters would equal number of subgenomes. By default, the number of clusters is one greater than the number of subgenomes, and the algorithm attempts to get rid of the ambiguous cluster  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-g">
<code class="descname">-g</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gamma</code><code class="descclassname"> &lt;gamma&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Gamma hyperparameter for Spectral Clustering and BGMM models.  [default: 1.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-mst">
<code class="descname">-mst</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_span_tree</code><code class="descclassname"> &lt;min_span_tree&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-mst" title="Permalink to this definition">¶</a></dt>
<dd><p>Augment k nearest neighbors graph with minimum spanning tree.  [default: 0]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-clustergraph">
<h3>clusterGraph<a class="headerlink" href="#python-polycracker-py-clustergraph" title="Permalink to this headline">¶</a></h3>
<p>Plots nearest neighbors graph in html format and runs a physics simulation on the graph over a number of iterations.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py clusterGraph <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-clustergraph-sp">
<code class="descname">-sp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--sparse_matrix_file</code><code class="descclassname"> &lt;sparse_matrix_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-clustergraph-sp" title="Permalink to this definition">¶</a></dt>
<dd><p>Sparse nearest neighbors graph npz file.  [default: spectralGraph.npz]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-clustergraph-sf">
<code class="descname">-sf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scaffolds_file</code><code class="descclassname"> &lt;scaffolds_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-clustergraph-sf" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file containing scaffold/chunk names.  [default: scaffolds_connect.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-clustergraph-od">
<code class="descname">-od</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--out_dir</code><code class="descclassname"> &lt;out_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-clustergraph-od" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory to output final plots.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-clustergraph-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--layout</code><code class="descclassname"> &lt;layout&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-clustergraph-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Layout from which to plot graph.  [default: standard]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-clustergraph-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--positions_npy</code><code class="descclassname"> &lt;positions_npy&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-clustergraph-p" title="Permalink to this definition">¶</a></dt>
<dd><p>If standard layout, then use these data points to begin simulation.  [default: graphInitialPositions.npy]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-clustergraph-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--iteration</code><code class="descclassname"> &lt;iteration&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-clustergraph-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Can comma delimit the number of iterations you would like to simulate to. No comma for a single cycle.  [default: 0,1,2,3]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-clustergraph-b">
<code class="descname">-b</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bed_features_file</code><code class="descclassname"> &lt;bed_features_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-clustergraph-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional bed file in relation to nonchunked genome. Must include features in fourth column.  [default: xxx]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-cluster-exchange">
<h3>cluster_exchange<a class="headerlink" href="#python-polycracker-py-cluster-exchange" title="Permalink to this headline">¶</a></h3>
<p>Prior to subgenome Extraction, can choose to switch out clusters, change clusters under examination, judging from clustered PCA plot if incorrect ambiguous cluster was found.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py cluster_exchange <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-exchange-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--subgenome_folder</code><code class="descclassname"> &lt;subgenome_folder&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-exchange-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Subgenome folder where subgenome extraction is taking place. Parent of the bootstrap_0 folder.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-cluster-exchange-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--list_of_clusters</code><code class="descclassname"> &lt;list_of_clusters&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-cluster-exchange-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Comma delimited list of numbers corresponding to cluster names. E.g. 0,1,2 corresponds to subgenome_0,subgenome_1, and subgenome_2, and those would be extracted.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-color-trees">
<h3>color_trees<a class="headerlink" href="#python-polycracker-py-color-trees" title="Permalink to this headline">¶</a></h3>
<p>Color phylogenetic trees by progenitor of origin.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py color_trees <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-color-trees-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-color-trees-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory for computations of Trees.  [default: ./TE_subclass_analysis/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-color-trees-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--newick_in</code><code class="descclassname"> &lt;newick_in&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-color-trees-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Input tree, newick format.  [default: ./x.treefile]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-color-trees-rd">
<code class="descname">-rd</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--top_repeat_dict_pickle</code><code class="descclassname"> &lt;top_repeat_dict_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-color-trees-rd" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing dict of top informative subgenome differential TEs by subgenome.  [default: ./TE_cluster_analysis/informative_subgenome_TE_dict.p]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-comparesubgenomes-progenitors-v-extracted">
<h3>compareSubgenomes_progenitors_v_extracted<a class="headerlink" href="#python-polycracker-py-comparesubgenomes-progenitors-v-extracted" title="Permalink to this headline">¶</a></h3>
<p>Compares the results found from the progenitorMapping function to results from polyCRACKER. Outputs CompareSubgenomesToProgenitors* files in the output directory. Final_stats is a better alternative to this function.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py compareSubgenomes_progenitors_v_extracted 
    <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scaffolds_file</code><code class="descclassname"> &lt;scaffolds_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to scaffolds pickle file.  [default: scaffolds.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-b1">
<code class="descname">-b1</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bed1</code><code class="descclassname"> &lt;bed1&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-b1" title="Permalink to this definition">¶</a></dt>
<dd><p>Bed file containing bootstrapped results.  [default: bootstrap_results.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-b2">
<code class="descname">-b2</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bed2</code><code class="descclassname"> &lt;bed2&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-b2" title="Permalink to this definition">¶</a></dt>
<dd><p>Bed file containing progenitor results. Alternatively, you can include the results from a prior bootstrap.  [default: progenitors.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-od">
<code class="descname">-od</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--out_dir</code><code class="descclassname"> &lt;out_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-comparesubgenomes-progenitors-v-extracted-od" title="Permalink to this definition">¶</a></dt>
<dd><p>Write results to this output directory.  [default: ./]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-compare-subclasses">
<h3>compare_subclasses<a class="headerlink" href="#python-polycracker-py-compare-subclasses" title="Permalink to this headline">¶</a></h3>
<p>In development: Grab abundance of top classes and subclasses of TEs in each subgenome and compare vs progenitors. Useful to use pairwise similarity function of all consensus repeats.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py compare_subclasses <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-compare-subclasses-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--subclass_dir</code><code class="descclassname"> &lt;subclass_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-compare-subclasses-s" title="Permalink to this definition">¶</a></dt>
<dd><p>[default: ./]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-convert-mat2r">
<h3>convert_mat2r<a class="headerlink" href="#python-polycracker-py-convert-mat2r" title="Permalink to this headline">¶</a></h3>
<p>Convert any sparse matrix into a format to be read by R. Can import matrix into R metagenomics clustering programs.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py convert_mat2r <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-convert-mat2r-npz">
<code class="descname">-npz</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_matrix</code><code class="descclassname"> &lt;input_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-convert-mat2r-npz" title="Permalink to this definition">¶</a></dt>
<dd><p>Input sparse matrix, scipy sparse npz format.  [default: clusteringMatrix.npz]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-convert-subgenome-output-to-pickle">
<h3>convert_subgenome_output_to_pickle<a class="headerlink" href="#python-polycracker-py-convert-subgenome-output-to-pickle" title="Permalink to this headline">¶</a></h3>
<p>Find cluster labels for all clusters/subgenomes in directory.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py convert_subgenome_output_to_pickle <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-id">
<code class="descname">-id</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_dir</code><code class="descclassname"> &lt;input_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-id" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing solely the cluster/classified outputs, with names stored in txt files. Can input bed file here as well.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scaffolds_pickle</code><code class="descclassname"> &lt;scaffolds_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to scaffolds pickle file.  [default: scaffolds.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_pickle</code><code class="descclassname"> &lt;output_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-convert-subgenome-output-to-pickle-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Desired output pickle name. Can feed colors_pickle.p, or name you choose, into plotPositions.  [default: colors_pickle.p]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-correct-kappa">
<h3>correct_kappa<a class="headerlink" href="#python-polycracker-py-correct-kappa" title="Permalink to this headline">¶</a></h3>
<p>Find corrected cohen’s kappa score. Used to test final_stats.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py correct_kappa <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-correct-kappa-dict">
<code class="descname">-dict</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--comparisons_dict</code><code class="descclassname"> &lt;comparisons_dict&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-correct-kappa-dict" title="Permalink to this definition">¶</a></dt>
<dd><p>MANDATORY: Comma delimited mapping of inferred subgenome names output from polycracker to Progenitor/Species labels.  [default: subgenome_1:progenitorA,subgenome_2:progenitorB,subgenome_3:speciesA]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-correct-kappa-p0">
<code class="descname">-p0</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scaffolds_pickle</code><code class="descclassname"> &lt;scaffolds_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-correct-kappa-p0" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file containing the original scaffold names.  [default: scaffolds_stats.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-correct-kappa-p1">
<code class="descname">-p1</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--polycracker_pickle</code><code class="descclassname"> &lt;polycracker_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-correct-kappa-p1" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file generated from final_stats. These are polycracker results.  [default: scaffolds_stats.poly.labels.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-correct-kappa-p2">
<code class="descname">-p2</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--progenitors_pickle</code><code class="descclassname"> &lt;progenitors_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-correct-kappa-p2" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file generated from final_stats. These are progenitor mapping results.  [default: scaffolds_stats.progenitors.labels.p]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-dash-genetic-algorithm-hdbscan-test">
<h3>dash_genetic_algorithm_hdbscan_test<a class="headerlink" href="#python-polycracker-py-dash-genetic-algorithm-hdbscan-test" title="Permalink to this headline">¶</a></h3>
<p>Save labels of all hdbscan runs, generate distance matrix for input into dash app. Output distance matrix, dataframe of parameters and score, and matrix of labels that produce distance matrix, referenced by parameters</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py dash_genetic_algorithm_hdbscan_test <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for plotting kmer count matrix is done.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_count_matrix</code><code class="descclassname"> &lt;kmer_count_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmer count matrix.  [default: ./kmer_master_count_matrix.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--reduction_method</code><code class="descclassname"> &lt;reduction_method&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of dimensionality reduction technique to use.  [default: tsne]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_dimensions</code><code class="descclassname"> &lt;n_dimensions&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of dimensions to reduce to.  [default: 3]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kernel</code><code class="descclassname"> &lt;kernel&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Kernel for KPCA. Cosine is particularly effective.  [default: cosine]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--metric</code><code class="descclassname"> &lt;metric&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Distance metric used to compute distance matrix.  [default: cosine]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-c">
<code class="descname">-min_c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_cluster_size</code><code class="descclassname"> &lt;min_cluster_size&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum cluster size for hdbscan algorithm.  [default: 300]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-s">
<code class="descname">-min_s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_samples</code><code class="descclassname"> &lt;min_samples&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum number samples for hdbscan algorithm.  [default: 150]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-nn">
<code class="descname">-nn</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_neighbors</code><code class="descclassname"> &lt;n_neighbors&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-nn" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of nearest neighbors in generation of nearest neighbor graph.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-scan">
<code class="descname">-scan</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hyperparameter_scan</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-scan" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to conduct hyperparameter scan of best clustering parameters.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-n">
<code class="descname">-min_n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_number_clusters</code><code class="descclassname"> &lt;min_number_clusters&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-min-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum number of clusters to find if doing hyperparameter scan. Parameters that produce fewer clusters face heavy penalization.  [default: 3]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-low">
<code class="descname">-low</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--low_counts</code><code class="descclassname"> &lt;low_counts&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-low" title="Permalink to this definition">¶</a></dt>
<dd><p>Comma delimited list of low bound on min_cluster_size, min_samples, and nearest neighbors respectively for hyperparameter scan.  [default: 5,5,5]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-j">
<code class="descname">-j</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_jobs</code><code class="descclassname"> &lt;n_jobs&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-j" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of jobs for TSNE transform.  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--silhouette</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Use mean silhouette coefficient with mahalanobis distance as scoring metric for GA hyperparameter scan.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-v">
<code class="descname">-v</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--validity</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genetic-algorithm-hdbscan-test-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Use hdbscan validity metric for clustering score. Takes precedent over silhouette score.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-dash-genome-quality-assessment">
<h3>dash_genome_quality_assessment<a class="headerlink" href="#python-polycracker-py-dash-genome-quality-assessment" title="Permalink to this headline">¶</a></h3>
<p>Input pre chromosome level scaffolded genomes, before split by ploidy, and outputs frequency distribution of select kmers.
Used to compare this distribution of kmers to that found from comparing post-scaffolded polyploid subgenomes.
The scaffolding, if reference based, may be biased and missing key repeat information, so comparing to prescaffolded genomes may show whether repeat info was lost during scaffolding for important words.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py dash_genome_quality_assessment <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genome-quality-assessment-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genome-quality-assessment-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for finding array of kmers versus prescaffolded genomes with unassigned sequences.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genome-quality-assessment-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_rule_matrix</code><code class="descclassname"> &lt;kmer_rule_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genome-quality-assessment-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmer count matrix with appended rule labels.  [default: ./kmer_master_matrix_rules.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genome-quality-assessment-g">
<code class="descname">-g</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--pre_scaffolded_genomes</code><code class="descclassname"> &lt;pre_scaffolded_genomes&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genome-quality-assessment-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Comma delimited dictionary of genome names and corresponding genomes. Genomes should be assemblies prior to polyCRACKER/bbtools binning and scaffolding.  [default: genome1:./path/genome_1_fname.fa,genome2:./path/genome_2_fname.fa,genome3:./path/genome_3_fname.fa]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-dash-genome-quality-assessment-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-dash-genome-quality-assessment-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-diff-kmer-analysis">
<h3>diff_kmer_analysis<a class="headerlink" href="#python-polycracker-py-diff-kmer-analysis" title="Permalink to this headline">¶</a></h3>
<p>Runs robust differential kmer analysis pipeline. Find highly informative differential kmers, subset of differential kmers found via polyCRACKER,
generate a matrix of highly informative differential kmer counts versus scaffolds, generate a network graph linking the kmers,
estimate phylogeny from these kmers, perform clustering using this matrix via a jaccard statistic, and look at the intersection between these kmers and identified repeats.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py diff_kmer_analysis <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for differential kmer analysis is done.  [default: ./diff_kmer_analysis/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-npz">
<code class="descname">-npz</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--sparse_kmer_matrix</code><code class="descclassname"> &lt;sparse_kmer_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-npz" title="Permalink to this definition">¶</a></dt>
<dd><p>Original sparse kmer matrix produced from generate_Kmer_Matrix  [default: clusteringMatrix.npz]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--final_output_label_pickle</code><code class="descclassname"> &lt;final_output_label_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle generated from bootstrap results from subgenomeExtraction. May need to run convert_subgenome_output_to_pickle.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-b">
<code class="descname">-b</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--final_output_label_bed</code><code class="descclassname"> &lt;final_output_label_bed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Bed generated from bootstrap results from subgenomeExtraction. May need to run generate_out_bed.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scaffolds_pickle</code><code class="descclassname"> &lt;scaffolds_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Scaffolds pickle produced from generate_Kmer_Matrix.  [default: scaffolds.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-kp">
<code class="descname">-kp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmers_pickle</code><code class="descclassname"> &lt;kmers_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-kp" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmers pickle produced from generate_Kmer_Matrix.  [default: kmers.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-min">
<code class="descname">-min</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_number_diff_kmers</code><code class="descclassname"> &lt;min_number_diff_kmers&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-min" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum number of differential kmers from each subgenome.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-dd">
<code class="descname">-dd</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--diff_kmers_directory</code><code class="descclassname"> &lt;diff_kmers_directory&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-dd" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing differential kmers from subgenomeExtraction bootstrap</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-go">
<code class="descname">-go</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_genome</code><code class="descclassname"> &lt;original_genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-go" title="Permalink to this definition">¶</a></dt>
<dd><p>Filename of original polyploid fasta, prechunked.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-fp">
<code class="descname">-fp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fasta_path</code><code class="descclassname"> &lt;fasta_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-fp" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to original polyploid fasta, prechunked.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-sl">
<code class="descname">-sl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--split_length</code><code class="descclassname"> &lt;split_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-sl" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of intervals in bedgraph files.  [default: 75000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_chromosomes</code><code class="descclassname"> &lt;n_chromosomes&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Output unionbed plots of x largest chromosomes.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-n-subgenomes">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_subgenomes</code><code class="descclassname"> &lt;n_subgenomes&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-n-subgenomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of subgenomes for Jaccard clustering.  [default: 2]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-nns">
<code class="descname">-nns</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_neighbors_scaffolds</code><code class="descclassname"> &lt;n_neighbors_scaffolds&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-nns" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of nearest neighbors to use when generating nearest neighbor graph with scaffolds as datapoints.  [default: 25]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-nnk">
<code class="descname">-nnk</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_neighbors_kmers</code><code class="descclassname"> &lt;n_neighbors_kmers&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-nnk" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of nearest neighbors to use when generating nearest neighbor graph with kmers as datapoints.  [default: 15]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kernel</code><code class="descclassname"> &lt;kernel&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Kernel for KPCA. Cosine is particularly effective.  [default: cosine]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-wnn">
<code class="descname">-wnn</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--weights</code><code class="descclassname"> &lt;weights&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-wnn" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to weight k_neighbors graph for jaccard clustering and kmer graphs.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--repeat_masked_gff</code><code class="descclassname"> &lt;repeat_masked_gff&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Repeatmasked gff file, in gff or gff3 format. Used for intersecting kmer results with Transposable Elements to find TE classes for kmers and other analyses.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-ndb">
<code class="descname">-ndb</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--no_diff_blast</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-ndb" title="Permalink to this definition">¶</a></dt>
<dd><p>Flag to not run earlier part of differential kmer analysis. Saves time if already done.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-ns">
<code class="descname">-ns</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--no_sam2matrix</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-ns" title="Permalink to this definition">¶</a></dt>
<dd><p>Flag, if on, does not convert sam files into a clustering matrix.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-np">
<code class="descname">-np</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--no_run_phylogeny</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-np" title="Permalink to this definition">¶</a></dt>
<dd><p>Flag, if on, does not run phylogeny analysis</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-diff-kmer-analysis-nj">
<code class="descname">-nj</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--no_jaccard_clustering</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-diff-kmer-analysis-nj" title="Permalink to this definition">¶</a></dt>
<dd><p>Flag, if on, does not run jaccard clustering</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-differential-te-histogram">
<h3>differential_TE_histogram<a class="headerlink" href="#python-polycracker-py-differential-te-histogram" title="Permalink to this headline">¶</a></h3>
<p>Compare the ratio of hits of certain highly informative repeats between different found genomes.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py differential_TE_histogram <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-differential-te-histogram-npy">
<code class="descname">-npy</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--differential_repeat_subgenome_matrix</code><code class="descclassname"> &lt;differential_TE_subgenome_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-differential-te-histogram-npy" title="Permalink to this definition">¶</a></dt>
<dd><p>[default: ./TE_cluster_analysis/subgenome_differential_top_TEs.npy]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-explore-kmers">
<h3>explore_kmers<a class="headerlink" href="#python-polycracker-py-explore-kmers" title="Permalink to this headline">¶</a></h3>
<p>Perform dimensionality reduction on matrix of kmer counts versus genomes and cluster using a combination of genetic algorithm + density based clustering (hdbscan). Uncovers kmer conservation patterns/rules across different lines.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py explore_kmers <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for plotting kmer count matrix is done.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_count_matrix</code><code class="descclassname"> &lt;kmer_count_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmer count matrix.  [default: ./kmer_master_count_matrix.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--reduction_method</code><code class="descclassname"> &lt;reduction_method&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of dimensionality reduction technique to use.  [default: tsne]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_dimensions</code><code class="descclassname"> &lt;n_dimensions&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of dimensions to reduce to.  [default: 3]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kernel</code><code class="descclassname"> &lt;kernel&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Kernel for KPCA. Cosine is particularly effective.  [default: cosine]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--metric</code><code class="descclassname"> &lt;metric&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Distance metric used to compute distance matrix.  [default: cosine]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-min-c">
<code class="descname">-min_c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_cluster_size</code><code class="descclassname"> &lt;min_cluster_size&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-min-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum cluster size for hdbscan algorithm.  [default: 300]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-min-s">
<code class="descname">-min_s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_samples</code><code class="descclassname"> &lt;min_samples&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-min-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum number samples for hdbscan algorithm.  [default: 150]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-nn">
<code class="descname">-nn</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_neighbors</code><code class="descclassname"> &lt;n_neighbors&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-nn" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of nearest neighbors in generation of nearest neighbor graph.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-scan">
<code class="descname">-scan</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hyperparameter_scan</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-scan" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to conduct hyperparameter scan of best clustering parameters.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-min-n">
<code class="descname">-min_n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_number_clusters</code><code class="descclassname"> &lt;min_number_clusters&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-min-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum number of clusters to find if doing hyperparameter scan. Parameters that produce fewer clusters face heavy penalization.  [default: 3]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-low">
<code class="descname">-low</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--low_counts</code><code class="descclassname"> &lt;low_counts&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-low" title="Permalink to this definition">¶</a></dt>
<dd><p>Comma delimited list of low bound on min_cluster_size, min_samples, and nearest neighbors respectively for hyperparameter scan.  [default: 5,5,5]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-j">
<code class="descname">-j</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_jobs</code><code class="descclassname"> &lt;n_jobs&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-j" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of jobs for TSNE transform.  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--silhouette</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Use mean silhouette coefficient with mahalanobis distance as scoring metric for GA hyperparameter scan.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-explore-kmers-v">
<code class="descname">-v</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--validity</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-explore-kmers-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Use hdbscan validity metric for clustering score. Takes precedent over silhouette score.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-extract-sequences">
<h3>extract_sequences<a class="headerlink" href="#python-polycracker-py-extract-sequences" title="Permalink to this headline">¶</a></h3>
<p>Extract sequences from fasta file and move them to new files as specified.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py extract_sequences <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-extract-sequences-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--sequences_dict</code><code class="descclassname"> &lt;sequences_dict&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-extract-sequences-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary of sequences and which subgenome/genome to send them to.  [default: S:S1,S2,S3-D:D1,D3,D10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-extract-sequences-fi">
<code class="descname">-fi</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fasta_in</code><code class="descclassname"> &lt;fasta_in&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-extract-sequences-fi" title="Permalink to this definition">¶</a></dt>
<dd><p>Fasta input file.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-final-stats">
<h3>final_stats<a class="headerlink" href="#python-polycracker-py-final-stats" title="Permalink to this headline">¶</a></h3>
<p>Analyzes the accuracy and agreement between polyCRACKER and ground truth species labels or progenitor mapped labels. Also compares polyCRACKER output to sizes of known progenitors. In future, may attempt to find similarity etween polyCRACKER sequences and progenitors through mash/bbsketch.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py final_stats <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--progenitors</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--species</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-p" title="Permalink to this definition">¶</a></dt>
<dd><p>If running final_stats with progenitors without known subgenomes labelling, will compare progenitor results to polyCRACKER without ground truth. If separating species from sample or you already know subgenome labelling, –species will compare to ground truth.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-dict">
<code class="descname">-dict</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--comparisons_dict</code><code class="descclassname"> &lt;comparisons_dict&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-dict" title="Permalink to this definition">¶</a></dt>
<dd><p>MANDATORY: Comma delimited mapping of inferred subgenome names output from polycracker to Progenitor/Species labels.  [default: subgenome_1:progenitorA,subgenome_2:progenitorB,subgenome_3:speciesA]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-cbed">
<code class="descname">-cbed</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--correspondence_bed</code><code class="descclassname"> &lt;correspondence_bed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-cbed" title="Permalink to this definition">¶</a></dt>
<dd><p>Bed file containing the original scaffold names. Please check if this bed file contains all scaffolds, else generate file from splitFasta.  [default: correspondence.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-pbed">
<code class="descname">-pbed</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--polycracker_bed</code><code class="descclassname"> &lt;polycracker_bed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-pbed" title="Permalink to this definition">¶</a></dt>
<dd><p>Bootstrap/cluster run results bed generated from outputgenerate_out_bed after running polycracker. These are polycracker results.  [default: subgenomes.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-sbed">
<code class="descname">-sbed</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--progenitors_species_bed</code><code class="descclassname"> &lt;progenitors_species_bed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-sbed" title="Permalink to this definition">¶</a></dt>
<dd><p>Progenitors bed output from progenitorMapping or species bed output.  [default: progenitors.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cluster_analysis</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to perform analyses on clustering results. Classification analysis will be disabled for this run if this option is selected, but you may need to add data from the above options.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-cs">
<code class="descname">-cs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--cluster_scaffold</code><code class="descclassname"> &lt;cluster_scaffold&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-cs" title="Permalink to this definition">¶</a></dt>
<dd><p>Scaffolds pickle file used for cluster analysis.  [default: scaffolds.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-npy">
<code class="descname">-npy</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--positions_npy</code><code class="descclassname"> &lt;positions_npy&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-npy" title="Permalink to this definition">¶</a></dt>
<dd><p>Positions npy file if choosing to do clustering analysis. Will assess the validity of the clusters with a euclidean metric, so may not be as valid for non-convex sets. Leave blank if not running analysis.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-op">
<code class="descname">-op</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_progenitors</code><code class="descclassname"> &lt;original_progenitors&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-op" title="Permalink to this definition">¶</a></dt>
<dd><p>Dictionary of original progenitor names to their fasta files. Eg. progenitorA:./progenitors/progenitorA.fasta,progenitorR:./progenitorsR.fasta</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-final-stats-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--note</code><code class="descclassname"> &lt;note&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-final-stats-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Note to self that can be put into report.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-find-best-cluster-parameters">
<h3>find_best_cluster_parameters<a class="headerlink" href="#python-polycracker-py-find-best-cluster-parameters" title="Permalink to this headline">¶</a></h3>
<p>In development: Experimenting with genetic algorithm (GA). Use GA to find best cluster method / hyperparameters given ground truth. These hyperparameter scans can help us find out what is important in finding good clusters rather than trial and error.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py find_best_cluster_parameters <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-best-cluster-parameters-f">
<code class="descname">-f</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--file</code><code class="descclassname"> &lt;file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-best-cluster-parameters-f" title="Permalink to this definition">¶</a></dt>
<dd><p>Numpy .npy file containing positions after dimensionality reduction</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-best-cluster-parameters-y">
<code class="descname">-y</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--y_true_pickle</code><code class="descclassname"> &lt;y_true_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-best-cluster-parameters-y" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file containing ground truths.  [default: colors_pickle.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-best-cluster-parameters-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_subgenomes</code><code class="descclassname"> &lt;n_subgenomes&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-best-cluster-parameters-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of subgenomes  [default: 3]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-best-cluster-parameters-g">
<code class="descname">-g</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--generations_number</code><code class="descclassname"> &lt;generations_number&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-best-cluster-parameters-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of generations.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-best-cluster-parameters-gm">
<code class="descname">-gm</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gene_mutation_prob</code><code class="descclassname"> &lt;gene_mutation_prob&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-best-cluster-parameters-gm" title="Permalink to this definition">¶</a></dt>
<dd><p>Gene mutation probability.  [default: 0.45]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-best-cluster-parameters-gc">
<code class="descname">-gc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gene_crossover_prob</code><code class="descclassname"> &lt;gene_crossover_prob&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-best-cluster-parameters-gc" title="Permalink to this definition">¶</a></dt>
<dd><p>Gene crossover probability.  [default: 0.45]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-best-cluster-parameters-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--population_size</code><code class="descclassname"> &lt;population_size&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-best-cluster-parameters-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Population size.  [default: 250]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-find-denovo-repeats">
<h3>find_denovo_repeats<a class="headerlink" href="#python-polycracker-py-find-denovo-repeats" title="Permalink to this headline">¶</a></h3>
<p>Wrapper for repeat modeler. Runs repeatmodeler and repeat masker to ID repeats.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py find_denovo_repeats <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-denovo-repeats-sp">
<code class="descname">-sp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_species</code><code class="descclassname"> &lt;input_species&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-denovo-repeats-sp" title="Permalink to this definition">¶</a></dt>
<dd><p>Input species for naming of repeat database.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-denovo-repeats-fi">
<code class="descname">-fi</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_fasta</code><code class="descclassname"> &lt;input_fasta&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-denovo-repeats-fi" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original fasta file, nonchunked.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-denovo-repeats-od">
<code class="descname">-od</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--out_dir</code><code class="descclassname"> &lt;out_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-denovo-repeats-od" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory for computations of denovo repeats.  [default: ./denovo_repeats/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-denovo-repeats-rd">
<code class="descname">-rd</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--recover_dir</code><code class="descclassname"> &lt;recover_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-denovo-repeats-rd" title="Permalink to this definition">¶</a></dt>
<dd><p>Recover directory when using RepeatModeler  [default: ]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-find-rules">
<h3>find_rules<a class="headerlink" href="#python-polycracker-py-find-rules" title="Permalink to this headline">¶</a></h3>
<p>In development: Elucidate underlying rules of certain kmer conservation patterns.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py find_rules <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-rules-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-rules-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for plotting kmer count matrix is done.  [default: ./]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-find-rules2">
<h3>find_rules2<a class="headerlink" href="#python-polycracker-py-find-rules2" title="Permalink to this headline">¶</a></h3>
<p>In development: Elucidate underlying rules of certain kmer conservation patterns.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py find_rules2 <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-rules2-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-rules2-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for plotting kmer count matrix is done.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-find-rules2-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">-n_rules</code><code class="descclassname"> &lt;n_rules&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-find-rules2-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of rules to find.  [default: 10]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-generate-kmer-matrix">
<h3>generate_Kmer_Matrix<a class="headerlink" href="#python-polycracker-py-generate-kmer-matrix" title="Permalink to this headline">¶</a></h3>
<p>From blasted bed file, where kmers were blasted to the polyploid genome, now generate a sparse matrix that contains information pertaining to the distribution of kmers (columns) for a given region of the genome (rows)</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py generate_Kmer_Matrix <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmercount_path</code><code class="descclassname"> &lt;kmercount_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing kmer count files</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-g">
<code class="descname">-g</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--genome</code><code class="descclassname"> &lt;genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-g" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of chunked genome fasta</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--chunk_size</code><code class="descclassname"> &lt;chunk_size&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of chunk of split/chunked fasta file  [default: 75000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-mc">
<code class="descname">-mc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_chunk_size</code><code class="descclassname"> &lt;min_chunk_size&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-mc" title="Permalink to this definition">¶</a></dt>
<dd><p>If min_chunk_threshold and prefilter flag is on, use chunks with lengths greater than a certain size  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--remove_non_chunk</code><code class="descclassname"> &lt;remove_non_chunk&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-r" title="Permalink to this definition">¶</a></dt>
<dd><p>If min_chunk_threshold and prefilter flag is on, remove chunks that are not the same size as chunk_size  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_chunk_threshold</code><code class="descclassname"> &lt;min_chunk_threshold&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-t" title="Permalink to this definition">¶</a></dt>
<dd><p>If prefilter is on, remove chunks with lengths less than a certain threshold  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--low_memory</code><code class="descclassname"> &lt;low_memory&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Must use with prefilter in order for prefilter to work. Without prefilter, populates matrix using unmerged blasted bed (slow).  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--prefilter</code><code class="descclassname"> &lt;prefilter&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-p" title="Permalink to this definition">¶</a></dt>
<dd><p>If selected, will now add chunks that were removed during preprocessing.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-f">
<code class="descname">-f</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fasta_path</code><code class="descclassname"> &lt;fasta_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-f" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing chunked polyploids fasta file</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-gs">
<code class="descname">-gs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--genome_split_name</code><code class="descclassname"> &lt;genome_split_name&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-gs" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of chunked genome fasta</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-go">
<code class="descname">-go</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_genome</code><code class="descclassname"> &lt;original_genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-go" title="Permalink to this definition">¶</a></dt>
<dd><p>File name of original, prechunked genome fasta</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_bed_file</code><code class="descclassname"> &lt;input_bed_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Input bed file, not merged.  [default: blasted.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-im">
<code class="descname">-im</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_merged_bed_file</code><code class="descclassname"> &lt;input_merged_bed_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-im" title="Permalink to this definition">¶</a></dt>
<dd><p>Input bed file, merged.  [default: blasted_merged.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-npz">
<code class="descname">-npz</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_sparse_matrix</code><code class="descclassname"> &lt;output_sparse_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-npz" title="Permalink to this definition">¶</a></dt>
<dd><p>Output sparse matrix file, in npz format.  [default: clusteringMatrix.npz]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_kmer_file</code><code class="descclassname"> &lt;output_kmer_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Output kmer pickle file.  [default: kmers.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-kmer-matrix-tf">
<code class="descname">-tf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--tfidf</code><code class="descclassname"> &lt;tfidf&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-kmer-matrix-tf" title="Permalink to this definition">¶</a></dt>
<dd><p>If set to one, no sequence length normalization will be done.  [default: 0]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-generate-genome-signatures">
<h3>generate_genome_signatures<a class="headerlink" href="#python-polycracker-py-generate-genome-signatures" title="Permalink to this headline">¶</a></h3>
<p>Wrapper for sourmash. Generate genome signatures and corresponding distance matrices for scaffolds/fasta files.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py generate_genome_signatures <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-genome-signatures-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_files</code><code class="descclassname"> &lt;input_files&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-genome-signatures-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Input fasta files. Put input in quotations if specifying multiple files. Eg. ‘genomes/<em>.fasta’  [default: ‘genomes/</em>.fa’]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-genome-signatures-kl">
<code class="descname">-kl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_length</code><code class="descclassname"> &lt;kmer_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-genome-signatures-kl" title="Permalink to this definition">¶</a></dt>
<dd><p>Default kmer length for signature.  [default: 23]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-genome-signatures-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_file_prefix</code><code class="descclassname"> &lt;output_file_prefix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-genome-signatures-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file prefix, default outputs output.cmp.dnd and output.sig.  [default: output]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-genome-signatures-mf">
<code class="descname">-mf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--multi_fasta</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-genome-signatures-mf" title="Permalink to this definition">¶</a></dt>
<dd><p>Instead of using multiple fasta files, generate signatures based on sequences in a multifasta file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-genome-signatures-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scaled</code><code class="descclassname"> &lt;scaled&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-genome-signatures-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Scale factor for sourmash.  [default: 10000]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-generate-karyotype">
<h3>generate_karyotype<a class="headerlink" href="#python-polycracker-py-generate-karyotype" title="Permalink to this headline">¶</a></h3>
<p>Generate karyotype shinyCircos/omicCircos csv file.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py generate_karyotype <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-karyotype-go">
<code class="descname">-go</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_genome</code><code class="descclassname"> &lt;original_genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-karyotype-go" title="Permalink to this definition">¶</a></dt>
<dd><p>Filename of polyploid fasta; must be full path to original.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-karyotype-shiny">
<code class="descname">--shiny</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-karyotype-shiny" title="Permalink to this definition">¶</a></dt>
<dd><p>Export to ShinyCircos.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-generate-out-bed">
<h3>generate_out_bed<a class="headerlink" href="#python-polycracker-py-generate-out-bed" title="Permalink to this headline">¶</a></h3>
<p>Find cluster labels for all clusters/subgenomes in directory, output to bed.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py generate_out_bed <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-out-bed-id">
<code class="descname">-id</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_dir</code><code class="descclassname"> &lt;input_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-out-bed-id" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing solely the cluster/classified outputs, with names stored in txt files.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-out-bed-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original</code><code class="descclassname"> &lt;original&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-out-bed-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Select 1 if subgenomeExtraction was done based on original nonchunked genome.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-out-bed-fai">
<code class="descname">-fai</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fai_file</code><code class="descclassname"> &lt;fai_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-out-bed-fai" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original, nonchunked fai file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-out-bed-fo">
<code class="descname">-fo</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_fname</code><code class="descclassname"> &lt;output_fname&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-out-bed-fo" title="Permalink to this definition">¶</a></dt>
<dd><p>Desired output bed file.  [default: output.bed]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-generate-unionbed">
<h3>generate_unionbed<a class="headerlink" href="#python-polycracker-py-generate-unionbed" title="Permalink to this headline">¶</a></h3>
<p>Generate a unionbedgraph with intervals of a specified length based on bed3 files from subgenomeExtraction</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py generate_unionbed <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-unionbed-b3">
<code class="descname">-b3</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bed3_directory</code><code class="descclassname"> &lt;bed3_directory&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-unionbed-b3" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing bed3 files produced from subgenomeExtraction.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-unionbed-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original</code><code class="descclassname"> &lt;original&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-unionbed-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Select 1 if subgenomeExtraction was done based on original nonchunked genome.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-unionbed-sl">
<code class="descname">-sl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--split_length</code><code class="descclassname"> &lt;split_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-unionbed-sl" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of intervals in bedgraph files.  [default: 75000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-unionbed-fai">
<code class="descname">-fai</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fai_file</code><code class="descclassname"> &lt;fai_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-unionbed-fai" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original, nonchunked fai file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-generate-unionbed-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_folder</code><code class="descclassname"> &lt;work_folder&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-generate-unionbed-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Working directory for processing subgenome bedgraphs and unionbedgraph.  [default: ./]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-genomic-label-propagation">
<h3>genomic_label_propagation<a class="headerlink" href="#python-polycracker-py-genomic-label-propagation" title="Permalink to this headline">¶</a></h3>
<p>Extend polyCRACKER labels up and downstream of found labels. Only works on split scaffold naming.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py genomic_label_propagation <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-genomic-label-propagation-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_labels</code><code class="descclassname"> &lt;input_labels&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-genomic-label-propagation-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file containing the cluster/class each label/scaffold belongs to.  [default: colors_pickle.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-genomic-label-propagation-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scaffolds_pickle</code><code class="descclassname"> &lt;scaffolds_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-genomic-label-propagation-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to scaffolds pickle file.  [default: scaffolds.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-genomic-label-propagation-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_pickle</code><code class="descclassname"> &lt;output_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-genomic-label-propagation-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file containing the new cluster/class of scaffolds.  [default: labels_new.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-genomic-label-propagation-npy">
<code class="descname">-npy</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--transformed_data</code><code class="descclassname"> &lt;transformed_data&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-genomic-label-propagation-npy" title="Permalink to this definition">¶</a></dt>
<dd><p>Input transformed matrix. Can leave empty, but if using this, label propagater will calculate distances from this matrix. Recommended use: differential kmer or TE matrix or PCA reduced data.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-genomic-label-propagation-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--radius</code><code class="descclassname"> &lt;radius&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-genomic-label-propagation-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Total radius in bps of search to look for neighboring classified sequences for RadiusNeighborsClassifier.  [default: 100000.0]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-get-density">
<h3>get_density<a class="headerlink" href="#python-polycracker-py-get-density" title="Permalink to this headline">¶</a></h3>
<p>Return gene or repeat density information in csv file from an input gff file. For shinyCircos.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py get_density <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-get-density-gff">
<code class="descname">-gff</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gff_file</code><code class="descclassname"> &lt;gff_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-get-density-gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Input gff file.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-get-density-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--window_length</code><code class="descclassname"> &lt;window_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-get-density-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Window length for histogram of gene density.  [default: 75000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-get-density-fai">
<code class="descname">-fai</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fai_file</code><code class="descclassname"> &lt;fai_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-get-density-fai" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original, nonchunked fai file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-get-density-fout">
<code class="descname">-fout</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--outputfname</code><code class="descclassname"> &lt;outputfname&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-get-density-fout" title="Permalink to this definition">¶</a></dt>
<dd><p>Output csv file name.  [default: gene_density.csv]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-hipmer-output-to-kcount">
<h3>hipmer_output_to_kcount<a class="headerlink" href="#python-polycracker-py-hipmer-output-to-kcount" title="Permalink to this headline">¶</a></h3>
<p>Converts hipmer kmer count output into a kmer count, kcount, file.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py hipmer_output_to_kcount <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-hipmer-output-to-kcount-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--hipmer_input</code><code class="descclassname"> &lt;hipmer_input&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-hipmer-output-to-kcount-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Input file or directory from hipmer kmer counting run.  [default: test.txt]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-hipmer-output-to-kcount-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kcount_output</code><code class="descclassname"> &lt;kcount_output&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-hipmer-output-to-kcount-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output kmer count file.  [default: test.final.kcount]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-hipmer-output-to-kcount-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--run_on_dir</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-hipmer-output-to-kcount-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Choose to run on all files in hipmer_input if you have specified a directory for the hipmer input. Directory can only contain hipmer files.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-kcount-hist">
<h3>kcount_hist<a class="headerlink" href="#python-polycracker-py-kcount-hist" title="Permalink to this headline">¶</a></h3>
<p>Outputs a histogram plot of a given kmer count files.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py kcount_hist <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kcount_directory</code><code class="descclassname"> &lt;kcount_directory&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing input kmer count files. Rename the kcount files to [insert-kmer-length].kcount, eg. 31.kcount . ln -s … may be useful in this case.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--out_file</code><code class="descclassname"> &lt;out_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output histogram in png.  [default: kmer_histogram.png]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-log">
<code class="descname">--log</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-log" title="Permalink to this definition">¶</a></dt>
<dd><p>Scale x-axis to log base 10.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kcount_range</code><code class="descclassname"> &lt;kcount_range&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Comma delimited list of two items, specifying range of kmer counts.  [default: 3,10000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--low_count</code><code class="descclassname"> &lt;low_count&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Input the kmer low count threshold and it will be plotted.  [default: 0.0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-kl">
<code class="descname">-kl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_lengths</code><code class="descclassname"> &lt;kmer_lengths&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-kl" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional: comma delimited list of kmer lengths to include in analysis. Eg. 31,54,73. Can optionally input the kmer low count corresponding to each length via another comma delimited list after colon, eg. Eg. 31,54,73:100,70,50 , where 100,70,50 are the low count thresholds</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--sample_speed</code><code class="descclassname"> &lt;sample_speed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional: increase sample speed to attempt to smooth kmer count histogram. A greater sample speed will help with smoothing more but sacrifice more information.  [default: 1]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-kcount-hist-old">
<h3>kcount_hist_old<a class="headerlink" href="#python-polycracker-py-kcount-hist-old" title="Permalink to this headline">¶</a></h3>
<p>Outputs a histogram plot of a given kmer count file.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py kcount_hist_old <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-old-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kcount_file</code><code class="descclassname"> &lt;kcount_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-old-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Input kmer count file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-old-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--out_file</code><code class="descclassname"> &lt;out_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-old-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output histogram in png or pdf.  [default: kmer_histogram.png]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kcount-hist-old-log">
<code class="descname">--log</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-kcount-hist-old-log" title="Permalink to this definition">¶</a></dt>
<dd><p>Scale x-axis to log base 10.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-kmer2fasta">
<h3>kmer2Fasta<a class="headerlink" href="#python-polycracker-py-kmer2fasta" title="Permalink to this headline">¶</a></h3>
<p>Converts kmer count file into a fasta file for blasting, after some threshold filtering and sampling</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py kmer2Fasta <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-kmer2fasta-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmercount_path</code><code class="descclassname"> &lt;kmercount_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kmer2fasta-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing kmer count files  [default: ./kmercount_files/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kmer2fasta-lc">
<code class="descname">-lc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_low_count</code><code class="descclassname"> &lt;kmer_low_count&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kmer2fasta-lc" title="Permalink to this definition">¶</a></dt>
<dd><p>Omit kmers from analysis that have less than x occurrences throughout genome.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kmer2fasta-hb">
<code class="descname">-hb</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--high_bool</code><code class="descclassname"> &lt;high_bool&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kmer2fasta-hb" title="Permalink to this definition">¶</a></dt>
<dd><p>Enter 1 if you would like to use the kmer_high_count option.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kmer2fasta-hc">
<code class="descname">-hc</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_high_count</code><code class="descclassname"> &lt;kmer_high_count&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kmer2fasta-hc" title="Permalink to this definition">¶</a></dt>
<dd><p>If high_bool is set to one, omit kmers that have greater than x occurrences.  [default: 2000000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-kmer2fasta-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--sampling_sensitivity</code><code class="descclassname"> &lt;sampling_sensitivity&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-kmer2fasta-s" title="Permalink to this definition">¶</a></dt>
<dd><p>If this option, x, is set greater than one, kmers are sampled at lower frequency, and total number of kmers included in the analysis is reduced to (total #)/(sampling_sensitivity), after threshold filtering  [default: 1]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-kmerratio2scaffasta">
<h3>kmerratio2scaffasta<a class="headerlink" href="#python-polycracker-py-kmerratio2scaffasta" title="Permalink to this headline">¶</a></h3>
<p>Bin genome regions into corresponding subgenomes based on the kmer counts in each region and their corresponding subgenomes</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py kmerratio2scaffasta <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="python-polycracker-py-link2color">
<h3>link2color<a class="headerlink" href="#python-polycracker-py-link2color" title="Permalink to this headline">¶</a></h3>
<p>Add color information to link file for shinyCircos.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py link2color <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-link2color-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--link_csv</code><code class="descclassname"> &lt;link_csv&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-link2color-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to link csv file.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-mash-test">
<h3>mash_test<a class="headerlink" href="#python-polycracker-py-mash-test" title="Permalink to this headline">¶</a></h3>
<p>Sourmash integration in development.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py mash_test <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="python-polycracker-py-merge-split-kmer-clusters">
<h3>merge_split_kmer_clusters<a class="headerlink" href="#python-polycracker-py-merge-split-kmer-clusters" title="Permalink to this headline">¶</a></h3>
<p>In development: working on merging and splitting clusters of particular kmer conservation patterns acros multiple genomes.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py merge_split_kmer_clusters <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-merge-split-kmer-clusters-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-merge-split-kmer-clusters-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for merging of clusters is done.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-merge-split-kmer-clusters-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--estimated_clusters</code><code class="descclassname"> &lt;estimated_clusters&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-merge-split-kmer-clusters-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Estimated number of clusters based on prior rule discovery. If set to -1, bases number of cluster selection on number already discovered.  [default: -1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-merge-split-kmer-clusters-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_dimensions</code><code class="descclassname"> &lt;n_dimensions&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-merge-split-kmer-clusters-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of dimensions to reduce to.  [default: 3]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-multicol2multifiles">
<h3>multicol2multifiles<a class="headerlink" href="#python-polycracker-py-multicol2multifiles" title="Permalink to this headline">¶</a></h3>
<p>Take matrix of total differential kmer counts, or similar matrix and break them up into single column files by found genome. For shinyCircos usage.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py multicol2multifiles <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-multicol2multifiles-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--multi_column_csv</code><code class="descclassname"> &lt;multi_column_csv&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-multicol2multifiles-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to multi-column csv file, likely containing list of differential kmer counts.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-out-bed-to-circos-csv">
<h3>out_bed_to_circos_csv<a class="headerlink" href="#python-polycracker-py-out-bed-to-circos-csv" title="Permalink to this headline">¶</a></h3>
<p>Take progenitor mapped, species ground truth, or polyCRACKER labelled scaffolds in bed file and convert for shinyCircos input of classification tracks..</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py out_bed_to_circos_csv <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-out-bed-to-circos-csv-b">
<code class="descname">-b</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_bed</code><code class="descclassname"> &lt;output_bed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-out-bed-to-circos-csv-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Output bed to convert into csv file for OmnicCircos plotting.  [default: subgenomes.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-out-bed-to-circos-csv-fai">
<code class="descname">-fai</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fai_file</code><code class="descclassname"> &lt;fai_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-out-bed-to-circos-csv-fai" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original, nonchunked fai file.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-plotpositions">
<h3>plotPositions<a class="headerlink" href="#python-polycracker-py-plotpositions" title="Permalink to this headline">¶</a></h3>
<p>Another plotting function without emphasis on plotting the spectral graph. Emphasis is on plotting positions and clusters.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py plotPositions <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-plotpositions-npy">
<code class="descname">-npy</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--positions_npy</code><code class="descclassname"> &lt;positions_npy&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plotpositions-npy" title="Permalink to this definition">¶</a></dt>
<dd><p>If standard layout, then use these data points to begin simulation.  [default: graphInitialPositions.npy]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plotpositions-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--labels_pickle</code><code class="descclassname"> &lt;labels_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plotpositions-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file containing scaffolds.  [default: scaffolds.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plotpositions-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--colors_pickle</code><code class="descclassname"> &lt;colors_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plotpositions-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file containing the cluster/class each label/scaffold belongs to.  [default: colors_pickle.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plotpositions-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_fname</code><code class="descclassname"> &lt;output_fname&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plotpositions-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Desired output plot name in html.  [default: output.html]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plotpositions-npz">
<code class="descname">-npz</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--graph_file</code><code class="descclassname"> &lt;graph_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plotpositions-npz" title="Permalink to this definition">¶</a></dt>
<dd><p>Sparse nearest neighbors graph npz file. If desired, try spectralGraph.npz.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plotpositions-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--layout</code><code class="descclassname"> &lt;layout&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plotpositions-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Layout from which to plot graph.  [default: standard]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plotpositions-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--iterations</code><code class="descclassname"> &lt;iterations&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plotpositions-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of iterations you would like to simulate to. No comma delimit, will only output a single iteration.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plotpositions-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--graph_sampling_speed</code><code class="descclassname"> &lt;graph_sampling_speed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plotpositions-s" title="Permalink to this definition">¶</a></dt>
<dd><p>When exporting the graph edges to CSV, can choose to decrease the number of edges for pdf report generation.  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plotpositions-ax">
<code class="descname">-ax</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--axes_off</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-plotpositions-ax" title="Permalink to this definition">¶</a></dt>
<dd><p>When enabled, exports graph without the axes.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-plot-distance-matrix">
<h3>plot_distance_matrix<a class="headerlink" href="#python-polycracker-py-plot-distance-matrix" title="Permalink to this headline">¶</a></h3>
<p>Perform dimensionality reduction on sourmash distance matrix or feature space matrix and plot samples in lower dimensional space and cluster.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py plot_distance_matrix <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-distance-matrix-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--dist_mat</code><code class="descclassname"> &lt;dist_mat&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-distance-matrix-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Input distance matrix from sourmash or can use feature matrix kmer_master_count_matrix.csv with option -f T.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-distance-matrix-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--transform_algorithm</code><code class="descclassname"> &lt;transform_algorithm&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-distance-matrix-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Dimensionality reduction technique to use.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-distance-matrix-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_file_prefix</code><code class="descclassname"> &lt;output_file_prefix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-distance-matrix-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file prefix, default outputs output.heatmap.png and output.[transform_algorithm].html.  [default: output]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-distance-matrix-j">
<code class="descname">-j</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_jobs</code><code class="descclassname"> &lt;n_jobs&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-distance-matrix-j" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of jobs for distance matrix transformation.  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-distance-matrix-f">
<code class="descname">-f</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--feature_space</code><code class="descclassname"> &lt;feature_space&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-distance-matrix-f" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix is in feature space (n_samples x n_features or vice versa); y transforms via assuming matrix is n_samples x n_features and T assumes matrix is n_features x n_samples.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-distance-matrix-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmeans_clusters</code><code class="descclassname"> &lt;kmeans_clusters&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-distance-matrix-k" title="Permalink to this definition">¶</a></dt>
<dd><p>If number of chosen clusters is greater than 0, runs kmeans clustering with chosen number of clusters, and choose most representative samples of clusters.  [default: 0]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-plot-rules">
<h3>plot_rules<a class="headerlink" href="#python-polycracker-py-plot-rules" title="Permalink to this headline">¶</a></h3>
<p>Plots normalized frequency distribution of select kmers across multiple genomes of certain rules/patterns.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py plot_rules <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-rules-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-rules-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for plotting kmer count matrix is done.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-rules-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--rule_csv</code><code class="descclassname"> &lt;rule_csv&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-rules-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmer count matrix with appended rule labels. Default includes chi-squared values.  [default: ./kmer_master_count_chi2_matrix_rules.csv]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-plot-rules-chromosomes">
<h3>plot_rules_chromosomes<a class="headerlink" href="#python-polycracker-py-plot-rules-chromosomes" title="Permalink to this headline">¶</a></h3>
<p>Plot distribution of rules/conservation pattern kmers across the chromosomes.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py plot_rules_chromosomes <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-rules-chromosomes-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-rules-chromosomes-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory for computations of spatial plotting of kmer rules.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-rules-chromosomes-go">
<code class="descname">-go</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_genome</code><code class="descclassname"> &lt;original_genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-rules-chromosomes-go" title="Permalink to this definition">¶</a></dt>
<dd><p>Complete path to original, prechunked genome fasta</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-rules-chromosomes-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_rule_matrix</code><code class="descclassname"> &lt;kmer_rule_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-rules-chromosomes-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmer count matrix with appended rule labels. Default includes chi-squared values.  [default: ./kmer_master_count_chi2_matrix_rules.csv]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-rules-chromosomes-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-rules-chromosomes-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-rules-chromosomes-sl">
<code class="descname">-sl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--split_length</code><code class="descclassname"> &lt;split_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-rules-chromosomes-sl" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of intervals, bed files.  [default: 75000]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-plot-unionbed">
<h3>plot_unionbed<a class="headerlink" href="#python-polycracker-py-plot-unionbed" title="Permalink to this headline">¶</a></h3>
<p>Plot results of union bed file, the distribution of total differential kmer counts for each extracted subgenome as a function across the entire genome. Note that results cannot be plot if original is set to one.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py plot_unionbed <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-unionbed-u">
<code class="descname">-u</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--unionbed_file</code><code class="descclassname"> &lt;unionbed_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-unionbed-u" title="Permalink to this definition">¶</a></dt>
<dd><p>Unionbed input file from subgenomeExtraction process or from other intermediate analyses.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-unionbed-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--number_chromosomes</code><code class="descclassname"> &lt;number_chromosomes&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-unionbed-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Output plots of x largest chromosomes.  [default: 10]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-plot-unionbed-od">
<code class="descname">-od</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--out_folder</code><code class="descclassname"> &lt;out_folder&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-plot-unionbed-od" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for plots.  [default: ./]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-polyploid-diff-kmer-comparison">
<h3>polyploid_diff_kmer_comparison<a class="headerlink" href="#python-polycracker-py-polyploid-diff-kmer-comparison" title="Permalink to this headline">¶</a></h3>
<p>Compare highly informative differential kmers between many different polyCRACKER runs/polyploids. Find intersections between them.
In development: May be better to modify bio_hyp_class pipeline to handle this problem.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py polyploid_diff_kmer_comparison <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--analysis_path</code><code class="descclassname"> &lt;analysis_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory where computations for polyploid comparison is done.  [default: ./polyploid_comparison/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_list_path</code><code class="descclassname"> &lt;kmer_list_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing list of informative differential kmers for each polyploid, pickle files.  [default: ./polyploid_comparison/diff_kmer_lists/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_type_path</code><code class="descclassname"> &lt;kmer_type_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-polyploid-diff-kmer-comparison-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing dictionaries of distributions of kmer class/subclass type for each polyploid, pickle files.  [default: ./polyploid_comparison/diff_kmer_types/]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-progenitormapping">
<h3>progenitorMapping<a class="headerlink" href="#python-polycracker-py-progenitormapping" title="Permalink to this headline">¶</a></h3>
<p>Takes reference progenitor fasta files, and bins an input polyploid file according to progenitor. These results will be compared to polyCRACKER’s results.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py progenitorMapping <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-progenitormapping-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_fasta</code><code class="descclassname"> &lt;input_fasta&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-progenitormapping-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Complete path to input polyploid fasta file, original or chunked.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-progenitormapping-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--progenitor_fasta_folder</code><code class="descclassname"> &lt;progenitor_fasta_folder&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-progenitormapping-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Folder containing progenitor genomes.  [default: ./progenitors/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-progenitormapping-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-progenitormapping-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-progenitormapping-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original</code><code class="descclassname"> &lt;original&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-progenitormapping-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Select 1 if trying to extract subgenomes based on original nonchunked genome.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-progenitormapping-kl">
<code class="descname">-kl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_length</code><code class="descclassname"> &lt;kmer_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-progenitormapping-kl" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of kmers to use for subgenome extraction.  [default: 23]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-progenitormapping-fout">
<code class="descname">-fout</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--outputfname</code><code class="descclassname"> &lt;outputfname&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-progenitormapping-fout" title="Permalink to this definition">¶</a></dt>
<dd><p>Output progenitor name  [default: progenitors.bed]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-progenitormapping-pf">
<code class="descname">-pf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--progenitor_fastas</code><code class="descclassname"> &lt;progenitor_fastas&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-progenitormapping-pf" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional: in addition to the progenitor fasta folder, directly specify the filename of each progenitor fasta and how you would like each fasta to be labelled as. Comma delimited list of fasta:label pairing. Example usage: -pf subgenomeT.fa:T,subgenomeA.fa:A,subgenomeR.fa:R</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-progenitormapping-gff">
<code class="descname">-gff</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--gff_file</code><code class="descclassname"> &lt;gff_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-progenitormapping-gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional:If progenitor_fastas specified, you have the option to input gff file, which will be split up into progenitor specific gene bed files for future synteny run.  [default: ]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-progenitormapping-go">
<code class="descname">-go</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_genome</code><code class="descclassname"> &lt;original_genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-progenitormapping-go" title="Permalink to this definition">¶</a></dt>
<dd><p>If running synteny: Filename of polyploid fasta; must be full path to original. Use if input gff file. Need to install jcvi library to run  [default: ]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-repeat-subclass-analysis">
<h3>repeat_subclass_analysis<a class="headerlink" href="#python-polycracker-py-repeat-subclass-analysis" title="Permalink to this headline">¶</a></h3>
<p>Input repeat_fasta and find phylogenies of TE subclasses within. In development: Better analysis may be <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2912890/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2912890/</a>.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py repeat_subclass_analysis <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-w">
<code class="descname">-w</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--work_dir</code><code class="descclassname"> &lt;work_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-w" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory for computations of TE_subclass_analysis.  [default: ./TE_subclass_analysis/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-go">
<code class="descname">-go</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_fasta</code><code class="descclassname"> &lt;original_fasta&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-go" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original fasta file, nonchunked.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-rp">
<code class="descname">-rp</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--top_repeat_list_pickle</code><code class="descclassname"> &lt;top_repeat_list_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-rp" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing list of top informative subgenome differential TEs.  [default: ./TE_cluster_analysis/informative_subgenome_TE_list.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-rd">
<code class="descname">-rd</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--top_repeat_dict_pickle</code><code class="descclassname"> &lt;top_repeat_dict_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-rd" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing dict of top informative subgenome differential TEs by subgenome.  [default: ./TE_cluster_analysis/informative_subgenome_TE_dict.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--all_family_subclass_pickle</code><code class="descclassname"> &lt;all_family_subclass_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle containing set of all TEs family, subclasses.  [default: ./TE_cluster_analysis/TE_motif_subclass.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-re">
<code class="descname">-re</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--all_repeat_elements</code><code class="descclassname"> &lt;all_repeat_elements&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-re" title="Permalink to this definition">¶</a></dt>
<dd><p>All repeat elements, to be converted to bed format.  [default: ./TE_cluster_analysis/repeat_elements.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-rs">
<code class="descname">-rs</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--class_subclass_pickle</code><code class="descclassname"> &lt;class_subclass_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-rs" title="Permalink to this definition">¶</a></dt>
<dd><p>All repeat classes and subclasses, to use to find indices to extract bed intervals.  [default: ./TE_cluster_analysis/repeat_class_subclass.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-bo">
<code class="descname">-bo</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_subgenomes_bed</code><code class="descclassname"> &lt;original_subgenomes_bed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-bo" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to original bed file containing subgenome labelling. Will use this to label TEs by subgenome.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-rf">
<code class="descname">-rf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--repeat_fasta</code><code class="descclassname"> &lt;repeat_fasta&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-rf" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to repeat fasta file, output from repeat masker pipeline.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-repeat-subclass-analysis-bt">
<code class="descname">-bt</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bootstrap</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-repeat-subclass-analysis-bt" title="Permalink to this definition">¶</a></dt>
<dd><p>Bootstrap subclass trees generation.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-reset-cluster">
<h3>reset_cluster<a class="headerlink" href="#python-polycracker-py-reset-cluster" title="Permalink to this headline">¶</a></h3>
<p>Delete cluster results, subgenome extraction results and corresponding html files.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py reset_cluster <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="python-polycracker-py-reset-transform">
<h3>reset_transform<a class="headerlink" href="#python-polycracker-py-reset-transform" title="Permalink to this headline">¶</a></h3>
<p>Remove all html files from main work directory. Must do this if hoping to retransform sparse data.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py reset_transform <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="python-polycracker-py-return-dash-data-structures">
<h3>return_dash_data_structures<a class="headerlink" href="#python-polycracker-py-return-dash-data-structures" title="Permalink to this headline">¶</a></h3>
<p>Return dash data structures needed to run dash app. Copy them over to a directory to be referenced by app or uploaded to Heroku for dash app deployment.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py return_dash_data_structures <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-return-dash-data-structures-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-return-dash-data-structures-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-return-dash-data-structures-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--reference_genome</code><code class="descclassname"> &lt;reference_genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-return-dash-data-structures-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Genome to blast against.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-return-dash-data-structures-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_dir</code><code class="descclassname"> &lt;output_dir&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-return-dash-data-structures-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Work directory for computations and outputs.  [default: ./]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-return-dash-data-structures-sl">
<code class="descname">-sl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--split_length</code><code class="descclassname"> &lt;split_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-return-dash-data-structures-sl" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of intervals in bedgraph files.  [default: 75000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-return-dash-data-structures-npy">
<code class="descname">-npy</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_pca_data</code><code class="descclassname"> &lt;kmer_pca_data&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-return-dash-data-structures-npy" title="Permalink to this definition">¶</a></dt>
<dd><p>Location of pca of kmer count matrix.  [default: ./kmer_counts_pca_results.npy]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-return-dash-data-structures-kl">
<code class="descname">-kl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_pca_labels</code><code class="descclassname"> &lt;kmer_pca_labels&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-return-dash-data-structures-kl" title="Permalink to this definition">¶</a></dt>
<dd><p>Location of kmer labels pickle for each pca datapoint, eg. ATTTCGGCGAT Bd: 10, Bs: 20, ABRD: 40.  [default: ./kmer_labels_coordinates.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-return-dash-data-structures-kr">
<code class="descname">-kr</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_rules_pickle</code><code class="descclassname"> &lt;kmer_rules_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-return-dash-data-structures-kr" title="Permalink to this definition">¶</a></dt>
<dd><p>Location of kmer rules pickle for each pca datapoint, eg.rule 3.  [default: ./rules.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-return-dash-data-structures-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_rule_matrix</code><code class="descclassname"> &lt;kmer_rule_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-return-dash-data-structures-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Kmer count matrix with appended rule labels. Default includes chi-squared values.  [default: ./kmer_master_count_chi2_matrix_rules.csv]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-run-iqtree">
<h3>run_iqtree<a class="headerlink" href="#python-polycracker-py-run-iqtree" title="Permalink to this headline">¶</a></h3>
<p>Perform multiple sequence alignment on multi-fasta and run iqtree to find phylogeny between each sequence.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py run_iqtree <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-iqtree-f">
<code class="descname">-f</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fasta_in</code><code class="descclassname"> &lt;fasta_in&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-iqtree-f" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to subclass repeat fasta. To be aligned via muscle.  [default: ./TE_subclass_analysis/Unknown.fa]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-iqtree-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--model</code><code class="descclassname"> &lt;model&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-iqtree-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Model selection for iqtree. See <a class="reference external" href="http://www.iqtree.org/doc/Substitution-Models">http://www.iqtree.org/doc/Substitution-Models</a>.  [default: MF]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-iqtree-nt">
<code class="descname">-nt</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_threads</code><code class="descclassname"> &lt;n_threads&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-iqtree-nt" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of threads for parallel computation.  [default: AUTO]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-iqtree-b">
<code class="descname">-b</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bootstrap</code><code class="descclassname"> &lt;bootstrap&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-iqtree-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Bootstrap support.  [default: 25]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-run-pipeline">
<h3>run_pipeline<a class="headerlink" href="#python-polycracker-py-run-pipeline" title="Permalink to this headline">¶</a></h3>
<p>Run polyCRACKER pipeline locally or on the cluster.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py run_pipeline <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-pipeline-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--submission_system</code><code class="descclassname"> &lt;submission_system&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-pipeline-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Job submission system to use, whether running jobs on sge, slurm, or locally. The configuration file will still be read, and nextflow jobs will still be submitted according to that specified on the config.  [default: local]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-pipeline-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--memory</code><code class="descclassname"> &lt;memory&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-pipeline-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use on sge or slurm submission job, in Gigabytes.  [default: 247]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-pipeline-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--time</code><code class="descclassname"> &lt;time&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-pipeline-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of hours to run sge or slurm submission job  [default: 11]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-pipeline-n">
<code class="descname">-N</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--nodes</code><code class="descclassname"> &lt;nodes&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-pipeline-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of nodes for slurm submission job  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-pipeline-b">
<code class="descname">-b</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--batch_script</code><code class="descclassname"> &lt;batch_script&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-pipeline-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Run custom batch script. Default will run the pipeline.  [default: runCluster.sh]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-pipeline-j">
<code class="descname">-J</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--job_name</code><code class="descclassname"> &lt;job_name&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-pipeline-j" title="Permalink to this definition">¶</a></dt>
<dd><p>Custom name for job.  [default: polyCRACKER]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-pipeline-no-options">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--no_options</code><code class="descclassname"></code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-pipeline-no-options" title="Permalink to this definition">¶</a></dt>
<dd><p>Run denovo slurm submission script with no options.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-pipeline-a">
<code class="descname">-a</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--account_name</code><code class="descclassname"> &lt;account_name&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-pipeline-a" title="Permalink to this definition">¶</a></dt>
<dd><p>Account name for running jobs on Cori.  [default: fungalp]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-run-pipeline-env">
<code class="descname">-env</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--environment</code><code class="descclassname"> &lt;environment&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-run-pipeline-env" title="Permalink to this definition">¶</a></dt>
<dd><p>Conda environment name to load when running polyCRACKER.  [default: polyCRACKER]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-run-tests">
<h3>run_tests<a class="headerlink" href="#python-polycracker-py-run-tests" title="Permalink to this headline">¶</a></h3>
<p>Run basic polyCRACKER tests to see if working properly.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py run_tests <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="python-polycracker-py-scaffolds2colors-specified">
<h3>scaffolds2colors_specified<a class="headerlink" href="#python-polycracker-py-scaffolds2colors-specified" title="Permalink to this headline">¶</a></h3>
<p>Attach labels to each scaffold for use of plotPositions. Colors scaffolds based on text within each scaffold. Useful for testing/evaluation purposes.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py scaffolds2colors_specified <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-scaffolds2colors-specified-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scaffolds_pickle</code><code class="descclassname"> &lt;scaffolds_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-scaffolds2colors-specified-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to scaffolds pickle file.  [default: scaffolds.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-scaffolds2colors-specified-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--labels_dict</code><code class="descclassname"> &lt;labels_dict&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-scaffolds2colors-specified-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Comma delimited dictionary of text to find in scaffold to new label. Example: text1:label1,text2:label2</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-scaffolds2colors-specified-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_pickle</code><code class="descclassname"> &lt;output_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-scaffolds2colors-specified-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Pickle file containing the new scaffold labels.  [default: colors_pickle.p]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-send-repeats">
<h3>send_repeats<a class="headerlink" href="#python-polycracker-py-send-repeats" title="Permalink to this headline">¶</a></h3>
<p>Use bbSketch to send fasta file containing repeats to check it against different databases for sequence similarity. Uses minhash to check relative abundance of species.
Where do these repeats come from and what is the parasitic DNA source.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py send_repeats <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-send-repeats-rf">
<code class="descname">-rf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--subclass_repeat_fasta</code><code class="descclassname"> &lt;subclass_repeat_fasta&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-send-repeats-rf" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to subclass repeat fasta.  [default: ./TE_subclass_analysis/Unknown.fa]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-shiny2omic">
<h3>shiny2omic<a class="headerlink" href="#python-polycracker-py-shiny2omic" title="Permalink to this headline">¶</a></h3>
<p>Convert shinyCircos csv input files to omicCircos csv input files.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py shiny2omic <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-shiny2omic-csv">
<code class="descname">-csv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--input_csv</code><code class="descclassname"> &lt;input_csv&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-shiny2omic-csv" title="Permalink to this definition">¶</a></dt>
<dd><p>Filename of shinyCircos csv.  [default: ]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-species-comparison-scaffold2colors">
<h3>species_comparison_scaffold2colors<a class="headerlink" href="#python-polycracker-py-species-comparison-scaffold2colors" title="Permalink to this headline">¶</a></h3>
<p>Generate color pickle file for plotPositions if list of scaffolds contains progenitors/species of origin. Useful for testing purposes.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py species_comparison_scaffold2colors <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-species-comparison-scaffold2colors-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--scaffolds_pickle</code><code class="descclassname"> &lt;scaffolds_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-species-comparison-scaffold2colors-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to scaffolds pickle file.  [default: scaffolds.p]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-species-comparison-scaffold2colors-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_pickle</code><code class="descclassname"> &lt;output_pickle&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-species-comparison-scaffold2colors-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to output species pickle file.  [default: colors_pickle.p]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-spectral-embed-plot">
<h3>spectral_embed_plot<a class="headerlink" href="#python-polycracker-py-spectral-embed-plot" title="Permalink to this headline">¶</a></h3>
<p>Spectrally embed PCA data of any origin.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py spectral_embed_plot <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-spectral-embed-plot-npy">
<code class="descname">-npy</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--positions</code><code class="descclassname"> &lt;positions&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-spectral-embed-plot-npy" title="Permalink to this definition">¶</a></dt>
<dd><p>Transformed data input, npy file.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-splitfasta">
<h3>splitFasta<a class="headerlink" href="#python-polycracker-py-splitfasta" title="Permalink to this headline">¶</a></h3>
<p>Split fasta file into chunks of a specified length</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py splitFasta <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-splitfasta-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fasta_file</code><code class="descclassname"> &lt;fasta_file&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-splitfasta-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Input polyploid fasta filename.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-splitfasta-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fasta_path</code><code class="descclassname"> &lt;fasta_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-splitfasta-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing polyploids fasta file  [default: ./fasta_files/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-splitfasta-c">
<code class="descname">-c</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--chunk_size</code><code class="descclassname"> &lt;chunk_size&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-splitfasta-c" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of chunk of split fasta file  [default: 75000]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-splitfasta-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--prefilter</code><code class="descclassname"> &lt;prefilter&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-splitfasta-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove chunks based on certain conditions as a preprocessing step to reduce memory  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-splitfasta-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_chunk_size</code><code class="descclassname"> &lt;min_chunk_size&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-splitfasta-m" title="Permalink to this definition">¶</a></dt>
<dd><p>If min_chunk_threshold and prefilter flag is on, use chunks with lengths greater than a certain size  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-splitfasta-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--remove_non_chunk</code><code class="descclassname"> &lt;remove_non_chunk&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-splitfasta-r" title="Permalink to this definition">¶</a></dt>
<dd><p>If min_chunk_threshold and prefilter flag is on, remove chunks that are not the same size as chunk_size  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-splitfasta-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--min_chunk_threshold</code><code class="descclassname"> &lt;min_chunk_threshold&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-splitfasta-t" title="Permalink to this definition">¶</a></dt>
<dd><p>If prefilter is on, remove chunks with lengths less than a certain threshold  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-splitfasta-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--low_memory</code><code class="descclassname"> &lt;low_memory&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-splitfasta-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Must use with prefilter in order for prefilter to work.  [default: 0]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-subgenomeextraction">
<h3>subgenomeExtraction<a class="headerlink" href="#python-polycracker-py-subgenomeextraction" title="Permalink to this headline">¶</a></h3>
<p>Extract subgenomes from genome, either chunked genome or original genome can be fed in as argument</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py subgenomeExtraction <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--subgenome_folder</code><code class="descclassname"> &lt;subgenome_folder&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Subgenome folder where subgenome extraction is taking place. This should be the bootstrap_0 folder.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-os">
<code class="descname">-os</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_subgenome_path</code><code class="descclassname"> &lt;original_subgenome_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-os" title="Permalink to this definition">¶</a></dt>
<dd><p>One level up from subgenomeFolder. Final outputs are stored here.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fasta_path</code><code class="descclassname"> &lt;fasta_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing chunked polyploids fasta file  [default: ./fasta_files/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-g">
<code class="descname">-g</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--genome_name</code><code class="descclassname"> &lt;genome_name&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Filename of chunked polyploid fasta.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-go">
<code class="descname">-go</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original_genome</code><code class="descclassname"> &lt;original_genome&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-go" title="Permalink to this definition">¶</a></dt>
<dd><p>Filename of polyploid fasta; can be either original or chunked, just make sure original is set to 1 if using original genome.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-bb">
<code class="descname">-bb</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bb</code><code class="descclassname"> &lt;bb&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-bb" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether bbtools were used in generating blasted sam file.  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-b">
<code class="descname">-b</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--bootstrap</code><code class="descclassname"> &lt;bootstrap&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of times to bootstrap the subgenome extraction process. Each time you bootstrap, you should get better classification results, else instability has occurred due to ambiguous clustering.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-i">
<code class="descname">-i</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--iteration</code><code class="descclassname"> &lt;iteration&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Current iteration of bootstrapping process. Please set to 0, when initially beginning.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_length</code><code class="descclassname"> &lt;kmer_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of kmers to find; can include multiple lengths if comma delimited (e.g. 23,25,27)  [default: 23,31]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-r">
<code class="descname">-r</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--run_final</code><code class="descclassname"> &lt;run_final&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Turn this to 0 when starting the command. Script terminates if this is set to 1.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--original</code><code class="descclassname"> &lt;original&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Select 1 if trying to extract subgenomes based on original nonchunked genome.  [default: 0]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-kl">
<code class="descname">-kl</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_low_count</code><code class="descclassname"> &lt;kmer_low_count&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-kl" title="Permalink to this definition">¶</a></dt>
<dd><p>Omit kmers from analysis that have less than x occurrences throughout genome.  [default: 100]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-dk">
<code class="descname">-dk</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--diff_kmer_threshold</code><code class="descclassname"> &lt;diff_kmer_threshold&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-dk" title="Permalink to this definition">¶</a></dt>
<dd><p>Value indicates that if the counts of a particular kmer in a subgenome divided by any of the counts in other subgenomes are greater than this value, then the kmer is differential.  [default: 20]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-u">
<code class="descname">-u</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--unionbed_threshold</code><code class="descclassname"> &lt;unionbed_threshold&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-u" title="Permalink to this definition">¶</a></dt>
<dd><p>Two comma delimited values used for binning regions into subgenomes. First value indicates that if the counts of all kmers in a particular region in any other subgenomes are 0, then the total count in a subgenome must be more than this much be binned as that subgenome. Second value indicates that if any of the other subgenome kmer counts for all kmers in a region are greater than 0, then the counts on this subgenome divided by any of the other counts must surpass a value to be binned as that subgenome.  [default: 10,2]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-ds">
<code class="descname">-ds</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--diff_sample_rate</code><code class="descclassname"> &lt;diff_sample_rate&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-ds" title="Permalink to this definition">¶</a></dt>
<dd><p>If this option, x, is set greater than one, differential kmers are sampled at lower frequency, and total number of differential kmers included in the analysis is reduced to (total #)/(diff_sample_rate), after threshold filtering  [default: 1]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-subgenomeextraction-kv">
<code class="descname">-kv</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--default_kmercount_value</code><code class="descclassname"> &lt;default_kmercount_value&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-subgenomeextraction-kv" title="Permalink to this definition">¶</a></dt>
<dd><p>If a particular kmer is not found in a kmercount dictionary, this number, x, will be used in its place. Useful for calculating differential kmers.  [default: 3.0]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-transform-plot">
<h3>transform_plot<a class="headerlink" href="#python-polycracker-py-transform-plot" title="Permalink to this headline">¶</a></h3>
<p>Perform dimensionality reduction on a sparse matrix and plot the results.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py transform_plot <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-transform-plot-t">
<code class="descname">-t</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--technique</code><code class="descclassname"> &lt;technique&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-transform-plot-t" title="Permalink to this definition">¶</a></dt>
<dd><p>Dimensionality reduction technique to use.  [default: kpca]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-transform-plot-s">
<code class="descname">-s</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_subgenomes</code><code class="descclassname"> &lt;n_subgenomes&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-transform-plot-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of subgenomes  [default: 2]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-transform-plot-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--metric</code><code class="descclassname"> &lt;metric&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-transform-plot-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Kernel for KPCA if kpca technique is chosen. Cosine is particularly effective.  [default: cosine]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-transform-plot-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--n_dimensions</code><code class="descclassname"> &lt;n_dimensions&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-transform-plot-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of dimensions to reduce to.  [default: 3]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-transform-plot-tf">
<code class="descname">-tf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--tfidf</code><code class="descclassname"> &lt;tfidf&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-transform-plot-tf" title="Permalink to this definition">¶</a></dt>
<dd><p>If set to one, tfidf normalization will be used instead of standard scaling.  [default: 0]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-txt2fasta">
<h3>txt2fasta<a class="headerlink" href="#python-polycracker-py-txt2fasta" title="Permalink to this headline">¶</a></h3>
<p>Extract subgenome fastas from reference fasta file using polyCRACKER found text files of subgenome binned scaffolds.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py txt2fasta <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-txt2fasta-txt">
<code class="descname">-txt</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--txt_files</code><code class="descclassname"> &lt;txt_files&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-txt2fasta-txt" title="Permalink to this definition">¶</a></dt>
<dd><p>Comma delimited list of text files or folder containing text files.  [default: subgenome_1.txt,subgenome_2.txt]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-txt2fasta-rf">
<code class="descname">-rf</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--reference_fasta</code><code class="descclassname"> &lt;reference_fasta&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-txt2fasta-rf" title="Permalink to this definition">¶</a></dt>
<dd><p>Full path to reference fasta file containing scaffold names.</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-unionbed2matrix">
<h3>unionbed2matrix<a class="headerlink" href="#python-polycracker-py-unionbed2matrix" title="Permalink to this headline">¶</a></h3>
<p>Convert unionbed file into a matrix of total differential kmer counts for each scaffold.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py unionbed2matrix <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-unionbed2matrix-u">
<code class="descname">-u</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--unionbed</code><code class="descclassname"> &lt;unionbed&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-unionbed2matrix-u" title="Permalink to this definition">¶</a></dt>
<dd><p>Unionbed file containing counts of differential kmers or TEs.  [default: subgenomes.union.bedgraph]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-unionbed2matrix-o">
<code class="descname">-o</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--output_matrix</code><code class="descclassname"> &lt;output_matrix&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-unionbed2matrix-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output path to differential kmer/TE matrix  [default: subgenomes.union.bed_matrix.npy]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-update-nextflow-config">
<h3>update_nextflow_config<a class="headerlink" href="#python-polycracker-py-update-nextflow-config" title="Permalink to this headline">¶</a></h3>
<p>Updates nextflow configuration file for use in nextflow pipeline.</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py update_nextflow_config <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-update-nextflow-config-p">
<code class="descname">-p</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--polycracker_config</code><code class="descclassname"> &lt;polycracker_config&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-update-nextflow-config-p" title="Permalink to this definition">¶</a></dt>
<dd><p>PolyCRACKER configuration file.  [default: config_polyCRACKER.txt]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-update-nextflow-config-n">
<code class="descname">-n</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--nextflow_config</code><code class="descclassname"> &lt;nextflow_config&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-update-nextflow-config-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Nextflow configuration file.  [default: nextflow.config]</p>
</dd></dl>

</div>
<div class="section" id="python-polycracker-py-writekmercount">
<h3>writeKmerCount<a class="headerlink" href="#python-polycracker-py-writekmercount" title="Permalink to this headline">¶</a></h3>
<p>Takes list of fasta files and runs kmercountexact.sh to find how many of each kmer of specified length is in genome</p>
<div class="highlight-shell"><div class="highlight"><pre><span></span>python polyCRACKER.py writeKmerCount <span class="o">[</span>OPTIONS<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="option">
<dt id="cmdoption-python-polycracker-py-writekmercount-d">
<code class="descname">-d</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--fasta_path</code><code class="descclassname"> &lt;fasta_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-writekmercount-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing chunked polyploids fasta file  [default: ./fasta_files/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-writekmercount-k">
<code class="descname">-k</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmercount_path</code><code class="descclassname"> &lt;kmercount_path&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-writekmercount-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing kmer count files  [default: ./kmercount_files/]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-writekmercount-l">
<code class="descname">-l</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--kmer_length</code><code class="descclassname"> &lt;kmer_length&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-writekmercount-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of kmers to find; can include multiple lengths if comma delimited (e.g. 23,25,27)  [default: 23,31]</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-python-polycracker-py-writekmercount-m">
<code class="descname">-m</code><code class="descclassname"></code><code class="descclassname">, </code><code class="descname">--blast_mem</code><code class="descclassname"> &lt;blast_mem&gt;</code><a class="headerlink" href="#cmdoption-python-polycracker-py-writekmercount-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Amount of memory to use for bbtools run. More memory will speed up processing.  [default: 100]</p>
</dd></dl>

</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Joshua Levy, Sean Gordon.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'v1.1.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>